{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"PrWazTVfeHMf"},"source":["#Lab.03 / IBM3202 – Phylogenetic Analysis using biopython and RAxML"]},{"cell_type":"markdown","metadata":{"id":"IA9WmdJLeQRm"},"source":["## Theoretical Aspects"]},{"cell_type":"markdown","metadata":{"id":"spFpaC5lenug"},"source":["Phylogenetic analysis lies at the core of genomics and bioinformatics and seeks to establish the evolutionary relationships between different homologous DNA or protein sequences and their ancestral sequences (common ancestors) from which they emerge. A typical result from a phylogenetic analysis is a **phylogenetic tree**, such as the following:"]},{"cell_type":"markdown","metadata":{"id":"Gh6BAOK51POj"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/phylo_01.png' width=500/>\n","<figcaption>FIGURE 1. A phylogenetic tree visually represents a hypothesis of how a group of sequences\n","are related. This figure explores how the way a tree is drawn conveys information. This tree shows how the seven sequences at the tips of the branches, called taxa, are related. Each horizontal branch represents an evolutionary\n","lineage. The length of the branch is arbitrary\n","unless the diagram specifies that branch lengths\n","represent information such as time or amount of\n","genetic change. Each branch point represents the common ancestor of the evolutionary lineages diverging from it. One of these branch points represents the common ancestor of all the sequences shown in this tree.\n","<br>Urry LA et al (2020). Campbell Biology. <i>Pearson Education, 12th Ed</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"lrMsCuYTe8Us"},"source":["In this representation, the tips of the tree (or the leaves) correspond to different, currently existing sequences of genes or proteins (usually called the **taxa**), thus they represent real data. The **nodes** or branch points connecting two sequences are the points of divergence in sequence evolution, namely **common ancestors** of extant taxa. The connection of two or more sequences with a hypothetical, extinct ancestral sequence allows grouping actual sequences into **clades**. On the other hand, the branch lengths represent the evolutionary changes between an ancestor and its descendant. Please notice that we are saying **'sequences'** and not **'species'**, as such statement requires the assumption of genetic isolation (i.e. all sequences in a given species show the same evolutionary drifts, but we learn from lectures that we can have horizontal gene transfer, right?)."]},{"cell_type":"markdown","metadata":{"id":"-dQoKNt5fNFO"},"source":["Many things can be said about this tree. First, a **phylogenetic tree** is based on a multiple sequence alignment and thus it is **as good as the alignment is**. See for example the polytomy in the ancestral node connecting taxa D, E and F; the sequence information between these sequences did not allow discerning how these sequences evolved from a common ancestor (i.e. imagine that there were two mutations and the phylogenetic analysis does not know which one comes first or second).\n","\n","Second, taxon G is defined as a **basal taxon**, a lineage that diverges from all other members of its group early in the history of the group.\n","\n","Lastly, the phylogenetic tree is **rooted** through an **outgroup**, i.e. a sequence (or a group or sequences) that is more distantly related to the ingroup sequences than the ingroup sequences are to each other; a distant homolog. As an outgroup sequence is not easy to define for all phylogenetic inferences, a vast number of phylogenetic trees in the literature are **unrooted** (i.e. there is no outgroup)."]},{"cell_type":"markdown","metadata":{"id":"hMLCidtMfd7E"},"source":["A phylogenetic tree is an estimation of the possible routes of evolution from ancestral lineages to present-day sequences that depends on many variables. We will smoothly go through all the steps of this inference, from sequence alignment to the “resurrection” of the most probable ancestral sequence."]},{"cell_type":"markdown","metadata":{"id":"5_sRCGKuf8Bf"},"source":["## Overview"]},{"cell_type":"markdown","metadata":{"id":"XMdhm8X_f_XN"},"source":["In this tutorial we will work with the family of forkhead box (Fox) transcription factors, a protein family that we currently study [Medina E et al (2016) *Biophys J 110 (11), 2349-2360*; Medina E et (2020) *J Mol Biol 432(19), 5411-5429*] in collaboration with the [Biochemistry and Molecular Biology Lab at U de Chile](https://sites.google.com/view/labbq) .\n","\n","These proteins are master controllers of gene expression in several eukaryotes since early *Bilateria*, and recently it has been determined that they also participate in chromosome remodeling since their structure is similar to the structure of histones. In humans, there are 19 different Fox subfamilies (from A to S), with almost all of them corresponding to monomeric proteins, except for FoxP. The members of the FoxP subfamily are known to form dimers to allow forming long-distance chromosome interactions due to a key Pro-to-Ala mutation in their sequence (see below)."]},{"cell_type":"markdown","metadata":{"id":"OHm2i0I071U3"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/phylo_02.png'/>\n","<figcaption>FIGURE 2. Sequence determinants of the structural acrobatics of the DNA-binding domain of human FoxP transcription factors. The left panel shows how FoxP proteins oligomerize by exchanging identical secondary structure elements between adjacent subunits, a mechanism known as domain swapping. An experimentally validated key Ala residue that enables domain swapping is indicated, which has been demonstrated to be conserved in all human FoxP proteins and replaced by Pro in all monomeric Fox members, as shown in the multiple sequence alignment on the right.\n","<br> Medina E et al (2016) <i>Biophys J 110 (11), 2349-2360</i>; Stroud JC et al (2008) <i>Structure 14(1), 159-66</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"QLCIaTFBgTcY"},"source":["Today, we will infer where did the evolutionary novelty of forming dimers emerged during the\n","evolution of Fox proteins using **RAxML** (Randomized Axelerated Maximum Likelihood), a popular program for phylogenetic analyses of large datasets under maximum likelihood."]},{"cell_type":"markdown","metadata":{"id":"buTRd47h4QfK"},"source":["#Part 0. Downloading and Installing the required software"]},{"cell_type":"markdown","metadata":{"id":"KSAoowL_4XIV"},"source":["Before we start, you must first **remember to start the hosted runtime in Google Colab**.\n","\n","Then, we must install several pieces of software to perform this tutorial. Namely:\n","- **biopython** for searching, retrieving, parsing and storing DNA and protein sequences.\n","- **MAFFT**, a multiple sequence alignment program for unix-like operating systems that offers a range of multiple alignment methods.\n","- **miniconda**, a free minimal installer of **conda** for software package and environment management.\n","- **ModelTest-ng**, a tool for selecting the best-fit model of evolution for DNA and protein alignments.\n","- **RAxML-ng**, a phylogenetic tree inference tool which uses maximum-likelihood (ML) optimality criterion.\n","\n","After several tests, the following installation instructions are the best way of setting up **Google Colab** for this laboratory session.\n","\n","1. We will first install biopython as follows:"]},{"cell_type":"code","metadata":{"id":"PdWvUiMR38gO"},"source":["#Installing biopython using pip\n","!pip install biopython"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"e3NwRC9A58AN"},"source":["#Importing biopython and os for safety\n","import os\n","import sys\n","import Bio"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"-6_wBK035ZW_"},"source":["2. Then, we will install MAFFT:"]},{"cell_type":"code","metadata":{"id":"H18SXwk5rikC"},"source":["#Installing MAFFT using apt-get in quiet mode\n","!apt-get install -qq -y mafft"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"umwoV7oR5-nH"},"source":["#Checking that MAFFT was succesfully installed\n","!mafft --help"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Goex7fGQTmcR"},"source":["3. Finally, we will install conda to be able to install RAxML-NG and ModelTest-NG as follows:"]},{"cell_type":"code","metadata":{"id":"i7x5g0HGxwpB"},"source":["#Install conda using the new conda-colab library\n","!pip install -q condacolab\n","import condacolab\n","condacolab.install_miniconda()\n","\n","#Install RAxML-NG and ModelTest-NG from\n","#the bioconda repository\n","!conda install -c bioconda raxml-ng modeltest-ng --yes\n","\n","#MAFFT could be also installed with conda\n","#using !conda install -c bioconda -y mafft\n","#We opted for apt-get instead"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"B4da3gT4hlfd"},"source":["#Part I - Retrieve Fox protein sequences using BLAST"]},{"cell_type":"markdown","metadata":{"id":"yBI5vVxLBss5"},"source":["Here, we will present two different options to retrieve Fox protein sequences.\n","1. Retrieving protein sequences using the web version of BLAST\n","2. Retrieving protein sequences using BLAST over biopython"]},{"cell_type":"markdown","metadata":{"id":"h2yIMMBtB9ml"},"source":["##Part I.A - Retrieve Fox protein sequences using the web version of BLAST"]},{"cell_type":"markdown","metadata":{"id":"KTbnS1RchwbY"},"source":["1. We will begin by opening a new window with [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi). Here, select **blastp** for protein sequences. Then, in the *“Enter accession number(s), gi(s) or FASTA sequence(s)”* box, write **2KIU_A**.\n","\n","  Also, we will indicate BLAST to only search for human sequences (**Search Set** option **Organism: humans**) from the Protein Data Bank (**Search Set** option **Database: Protein Data Bank (pdb)**), as shown below."]},{"cell_type":"markdown","metadata":{"id":"r1S9zsPcUBFi"},"source":["**QUESTION:** There are different types of BLAST. What does each one of them do?"]},{"cell_type":"markdown","metadata":{"id":"wWJ1jfAbM6QC"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/phylo_03.png'/>\n","</center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"9tzgc-4_jlgC"},"source":["2. The results from this BLAST are shown below. It is important to note that the first hit is actually our query sequence: **2KIU**. Also, all other sequences have similar 4-character codes.\n","\n","  If you remember from our previous tutorial on protein structure visualization using **py3Dmol**, these 4-character codes correspond to **Protein Data Bank (PDB)** accession IDs. These structures are always accompanied by a FASTA sequence of the crystallized protein.\n","\n","  We **strongly recommend** you to access the [PDB website](https://www.rcsb.org/) and analyze the information that is deposited for each one of these structures."]},{"cell_type":"markdown","metadata":{"id":"ktmXSPX1PXAP"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/phylo_04.png'/>\n","</center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"auv671TTDIoo"},"source":["3. We will download some of these sequences for our work. To collect these sequences for future analysis (e.g. multiple sequence alignments) we need to select the sequences that we would like to download, click on **Download** and select the **FASTA (complete sequences)** option. Then, you can either upload this file or copy these sequences onto a new file in Google Colab.\n"]},{"cell_type":"markdown","metadata":{"id":"SUCAF8HNkOn8"},"source":["**QUESTION:** Which parameters from BLAST would be important to decide which sequences to collect and which sequences to discard? Are we able to include more sequences?"]},{"cell_type":"markdown","metadata":{"id":"al-iSDTlk02T"},"source":["**💡 HINT:** Please note that we are using the PDB protein database only for the purposes of this tutorial, as it has a limited number of protein sequences (due to the limited number of structures). However, you might find the use of another database more suitable for your work with your own protein or DNA sequences, as they contain a higher number of sequences, hence a higher sequence variability/redundancy."]},{"cell_type":"markdown","metadata":{"id":"tCAjEWmahx3T"},"source":["## Part I.B - Retrieve Fox protein sequences using biopython"]},{"cell_type":"markdown","metadata":{"id":"uysnl6lrW2rb"},"source":["Biopython is an excellent companion for working with DNA and protein sequences and also with structures. Here, we will show how to use it for retrieving Fox protein sequences.\n","\n","1. First, we will start by using _Entrez_ to retrieve the sequence of accession code **2KIU_A** in FASTA format and _SeqIO_ to be able to read, parse and/or write this sequence. Besides downloading our sequence, we will also store the sequence into a string (here, termed as _aaseq_) for its subsequent use in this tutorial."]},{"cell_type":"code","metadata":{"id":"vt1ke8upXSl9"},"source":["from Bio import SeqIO, Entrez\n","#Setting up your email to be able to use Entrez\n","Entrez.email = 'your.email@uc.cl'\n","#Here, we set up a temporary handle with our downloaded sequence in fasta format\n","temp = Entrez.efetch(db=\"protein\",rettype=\"fasta\",id=\"2KIU_A\")\n","#Creating a fasta file to write our downloaded sequence\n","aaseq_out = open(\"2KIU_A.fasta\",'w')\n","#Reading the sequence information as a string in fasta format\n","aaseq = SeqIO.read(temp, format=\"fasta\")\n","#Writing the sequence record in fasta format\n","SeqIO.write(aaseq,aaseq_out,\"fasta\")\n","#Closing both the temp handle and the FASTA file\n","temp.close()\n","aaseq_out.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"klOzI65TXfJ6"},"source":["2. What is great about _SeqIO_ is that you can use it to manipulate your sequence (e.g. sorting, changing formats, etc) and also to print information about your sequence, such as its description, sequence and accession ID.\n","\n","  You can try these commands below by first writing **\"aaseq.\"** and then selecting one of the autocomplete options suggested by Google Colab. With this information, try to obtain the information requested below."]},{"cell_type":"code","metadata":{"id":"gy8YLQM3Z151"},"source":["#Printing the number of amino acids as an example\n","print(\"Sequence length (aa):\")\n","print(len(aaseq))"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"KjE6JtzTaXD1"},"source":["#Printing the description of the sequence\n","print(\"The description of the sequence is:\")\n","#INSERT YOUR OWN COMMAND HERE\n","print(aaseq.description)\n","\n","#Printing the accession ID of the sequence\n","print(\"The ID of the sequence is:\")\n","#INSERT YOUR OWN COMMAND BELOW\n","print(aaseq.id)\n","\n","#Printing the sequence\n","print(\"The sequence is:\")\n","#INSERT YOUR OWN COMMAND BELOW\n","print(aaseq.seq)\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"3bQo87E1lmfR"},"source":["**💡 HINT:** If you only have a protein/nucleic sequence instead of an accession ID, you can store the sequence in a string using the _**Seq**_ function as shown below."]},{"cell_type":"code","metadata":{"id":"yOJ6ou9lj0ef"},"source":["#In case you do not have an accession ID, but just a sequence\n","#This is not needed for our tutorial!\n","from Bio.Seq import Seq\n","my_seq = Seq(\"AEVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENVKGAVWTVDEVEFQKRRPQK\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"HnVcH7l4izF2"},"source":["3. Once we have already stored the information of our query sequence in a string, we can use it to perform a BLAST search inside Google Colab via biopython through _NCBIWWW_.\n","\n","  The code cell bellow specifies the blast program **blastp**, the database **pdb** and the query sequence. This process should not take longer than 2 min.\n","\n","  **💡 HINT:** Please note that we are using the PDB protein database only for the purposes of this tutorial, as it has a limited number of protein sequences (due to the limited number of structures). However, you might find the use of another database more suitable for your work with your own protein or DNA sequences, as they contain a higher number of sequences, hence a higher sequence variability/redundancy. For the purposes of this tutorial, we are also specifying to retrieve sequences only belonging to humans."]},{"cell_type":"code","metadata":{"id":"ChxitIMEkw8E"},"source":["#Using biopython to perform a blast search\n","%%time\n","from Bio.Blast import NCBIWWW\n","#NCBIWWW.qblast(program, database, sequence)\n","result_handle = NCBIWWW.qblast(\"blastp\", \"pdb\", aaseq.seq, entrez_query='human[organism]')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"DC3IwSuimbaW"},"source":["4. In order to parse this data, we need to store it in a handle for post-processing using _NCBIXML_, as shown below:"]},{"cell_type":"code","metadata":{"id":"MZozLVEcnfA1"},"source":["#Read the results in XML format for parsing BLAST records\n","from Bio.Blast import NCBIXML\n","blast_record = NCBIXML.read(result_handle)\n","#This is required to reset the searches and start over again\n","result_handle.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"V7qRGKKsn_UY"},"source":["5. _NCBIXML_ enables data manipulation similarly to what we saw before for _SeqIO_. In the following code cell, try to obtain the BLAST **hit id**, sequence **length** and **expect**-value or e-value for each high-scoring pair (HSP), using again the autocomplete function of Google Colab."]},{"cell_type":"code","metadata":{"id":"3hnxPKj-oqa1"},"source":["#Testing the functions of NCBIXML parsing\n","for alignment in blast_record.alignments:\n","  for hsp in alignment.hsps:\n","    #Example with the accession code\n","    print(\"Hit ID:\")\n","    print(alignment.hit_id)\n","    #Now try it yourself for printing the sequence lenght\n","\n","    print(\"Hit sequence length:\")\n","    #INSERT YOUR OWN COMMAND BELOW\n","    print(hsp.align_length)\n","\n","    #Now print the e-value for each hit\n","    print(\"E-value:\")\n","    #INSERT YOUR OWN COMMAND BELOW\n","    print(hsp.expect)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"VvWpHiceqNi5"},"source":["6. We can use all these functions to print a BLAST-like briefing of the results, and to even filter the results according to parameters such as the sequence identity, coverage and/or e-value. **Carefully examine the commands below to achieve this and the resulting output.**"]},{"cell_type":"code","metadata":{"id":"atHlk0A-iRtZ"},"source":["#Here, we show how we can set a sequence identity cut-off\n","#Many other cut-offs can be employed!\n","PIDcut=1.00\n","#Printing all BLAST parameters similarly to the website\n","for alignment in blast_record.alignments:\n","  for hsp in alignment.hsps:\n","#Here, we add a condition to only print hits equal or less than a PID cutoff\n","    if(hsp.identities/hsp.align_length) <= PIDcut:\n","      print(\"Accession code:\", alignment.hit_id)\n","      print(\"Sequence length:\", alignment.length)\n","      print(\"Alignment length:\", hsp.align_length)\n","      print(\"E-value:\", hsp.expect)\n","#The following command calculates the sequence identity relative to\n","#the length of the alignment\n","      print(\"Sequence Identity [%]:\", \"{:.2f}\".format(100*hsp.identities/hsp.align_length))\n","#The following command calculates the sequence coverage relative to\n","#the lenght of the sequence\n","      print(\"Sequence Coverage [%]:\", \"{:.2f}\".format(100*sum(c.isalpha() for c in hsp.query)/len(aaseq)))\n","      print()\n","#Here, we print the first 60 characters of the query and hit and their matches\n","      print(\"query:\", hsp.query[0:60] + \"...\")\n","      print(\"      \", hsp.match[0:60] + \"...\")\n","      print(\"sbjct:\", hsp.sbjct[0:60] + \"...\")\n","      print()\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"w_XmaOzfsYgJ"},"source":["**QUESTION:** Which parameters from BLAST would be important to decide which sequences to collect and which sequences to discard? Are we able to include more sequences?"]},{"cell_type":"markdown","metadata":{"id":"4cKEj6ApscwF"},"source":["**📚 HOMEWORK:** Copy the code cell above and edit it to add other filters, such as a minimum E-value and sequence coverage."]},{"cell_type":"markdown","metadata":{"id":"9IwBZmXOtmc5"},"source":["7. After the BLAST search and filtering, we would like to retrieve all sequences that match our selection criteria. Here, we opted to download the top 20 sequences, but you can use filters such as minimum sequence coverage, minimum e-value and maximum PID to limit your output sequences."]},{"cell_type":"code","metadata":{"id":"NWaDVmpp3JDx"},"source":["#Setting up your email to be able to use Entrez\n","Entrez.email = 'your.email@uc.cl'\n","\n","#Generate a loop to write all sequences into an output file\n","with open(\"sequences.fasta\", \"a\") as allhits_out:\n","#Check how we are indicating to use the top 20 hits\n","  for alignment in blast_record.alignments[:20]:\n","    for hsp in alignment.hsps:\n","    #Here, we add a condition to print only sequences below a PID cutoff\n","      if(hsp.identities/len(hsp.match)) <= PIDcut:\n","        print(\"Fetching protein sequence:\", alignment.hit_id)\n","        fetch = Entrez.efetch(db=\"protein\", id=alignment.hit_id, rettype=\"fasta\")\n","        #Reading the sequence stored in the temporary string in fasta format\n","        allhits_seqs = SeqIO.read(fetch, format=\"fasta\")\n","        #Writing the sequence and its ID in fasta format\n","        SeqIO.write(allhits_seqs,allhits_out,\"fasta\")\n","        fetch.close()\n","#Closing the efetch and file\n","allhits_out.close()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"7a8lnCG8xkEx"},"source":["😱 **EMERGENCY BACKUP!** In case BLASTP fails due to an issue with the NCBI servers, you can download a FASTA file containing 20 protein sequence homologs generated on Sep 27, 2021:"]},{"cell_type":"code","metadata":{"id":"1WWm3QqTyFgW"},"source":["!wget https://github.com/pb3lab/ibm3202/raw/master/files/emergency_backup/lab03/sequences.fasta -O sequences.fasta"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"9h_NN5COWGjo"},"source":["#Part II - Obtain and edit a Multiple Sequence Alignment (MSA) using MAFFT and biopython"]},{"cell_type":"markdown","metadata":{"id":"bEM7r3aL87xu"},"source":["Once BLAST is resolved, we can proceeed with the **Multiple Sequence Alignment (MSA)**. In this case, we will first employ **MAFFT** to align all retrieved sequences.\n","\n","1. To perform a MSA alignment using MAFFT, is very simple!"]},{"cell_type":"code","metadata":{"id":"ql6daJM9bMuS"},"source":["!mafft sequences.fasta > aligned.fasta"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"4_wbJAK9-Bku"},"source":["2. The result from this algorithm, which is an extension of the **Pairwise Alignment Algorithms** we discussed during Lectures, can be seen in an online MSA viewer such as [Alignment Viewer 2.0](https://fast.alignmentviewer.org/). Nevertheless, some recent developments using **Panel** widgets and the **Bokeh** interactive plotting library for Python for use in web browsers and dashboards. Here, we will be using one of such developments."]},{"cell_type":"code","metadata":{"id":"UYrebWipbp5A","cellView":"form"},"source":["#@title Protein MSA Viewer in Google Colab\n","#The following code is modified from the wonderful viewer developed by Damien Farrell\n","#https://dmnfarrell.github.io/bioinformatics/bokeh-sequence-aligner\n","\n","#Importing all modules first\n","import os, io, random\n","import string\n","import numpy as np\n","\n","from Bio.Seq import Seq\n","from Bio.Align import MultipleSeqAlignment\n","from Bio import AlignIO, SeqIO\n","\n","import panel as pn\n","import panel.widgets as pnw\n","pn.extension()\n","\n","from bokeh.plotting import figure\n","from bokeh.models import ColumnDataSource, Plot, Grid, Range1d\n","from bokeh.models.glyphs import Text, Rect\n","from bokeh.layouts import gridplot\n","\n","#Setting up the amino color code according to Zappo color scheme\n","def get_colors(seqs):\n","    #make colors for bases in sequence\n","    text = [i for s in list(seqs) for i in s]\n","    #Use Zappo color scheme\n","    clrs =  {'K':'red',\n","             'R':'red',\n","             'H':'red',\n","             'D':'green',\n","             'E':'green',\n","             'Q':'blue',\n","             'N':'blue',\n","             'S':'blue',\n","             'T':'blue',\n","             'A':'blue',\n","             'I':'blue',\n","             'L':'blue',\n","             'M':'blue',\n","             'V':'blue',\n","             'F':'orange',\n","             'Y':'orange',\n","             'W':'orange',\n","             'C':'blue',\n","             'P':'yellow',\n","             'G':'orange',\n","             '-':'white'}\n","    colors = [clrs[i] for i in text]\n","    return colors\n","\n","#Setting up the MSA viewer\n","def view_alignment(aln, fontsize=\"9pt\", plot_width=800):\n","    \"\"\"Bokeh sequence alignment view\"\"\"\n","\n","    #make sequence and id lists from the aln object\n","    seqs = [rec.seq for rec in (aln)]\n","    ids = [rec.id for rec in aln]\n","    text = [i for s in list(seqs) for i in s]\n","    colors = get_colors(seqs)\n","    N = len(seqs[0])\n","    S = len(seqs)\n","    width = .4\n","\n","    x = np.arange(1,N+1)\n","    y = np.arange(0,S,1)\n","    #creates a 2D grid of coords from the 1D arrays\n","    xx, yy = np.meshgrid(x, y)\n","    #flattens the arrays\n","    gx = xx.ravel()\n","    gy = yy.flatten()\n","    #use recty for rect coords with an offset\n","    recty = gy+.5\n","    h= 1/S\n","    #now we can create the ColumnDataSource with all the arrays\n","    source = ColumnDataSource(dict(x=gx, y=gy, recty=recty, text=text, colors=colors))\n","    plot_height = len(seqs)*15+50\n","    x_range = Range1d(0,N+1, bounds='auto')\n","    if N>100:\n","        viewlen=100\n","    else:\n","        viewlen=N\n","    #view_range is for the close up view\n","    view_range = (0,viewlen)\n","    tools=\"xpan, xwheel_zoom, reset, save\"\n","\n","    #entire sequence view (no text, with zoom)\n","    p = figure(title=None, width= plot_width, height=50,\n","               x_range=x_range, y_range=(0,S), tools=tools,\n","               min_border=0, toolbar_location='below')\n","    rects = Rect(x=\"x\", y=\"recty\",  width=1, height=1, fill_color=\"colors\",\n","                 line_color=None, fill_alpha=0.6)\n","    p.add_glyph(source, rects)\n","    p.yaxis.visible = False\n","    p.grid.visible = False\n","\n","    #sequence text view with ability to scroll along x axis\n","    p1 = figure(title=None, width=plot_width, height=plot_height,\n","                x_range=view_range, y_range=ids, tools=\"xpan,reset\",\n","                min_border=0, toolbar_location='below')#, lod_factor=1)\n","    glyph = Text(x=\"x\", y=\"y\", text=\"text\", text_align='center',text_color=\"black\",\n","                text_font=\"monospace\",text_font_size=fontsize)\n","    rects = Rect(x=\"x\", y=\"recty\",  width=1, height=1, fill_color=\"colors\",\n","                line_color=None, fill_alpha=0.4)\n","    p1.add_glyph(source, glyph)\n","    p1.add_glyph(source, rects)\n","\n","    p1.grid.visible = False\n","    p1.xaxis.major_label_text_font_style = \"bold\"\n","    p1.yaxis.minor_tick_line_width = 0\n","    p1.yaxis.major_tick_line_width = 0\n","\n","    p = gridplot([[p],[p1]], toolbar_location='below')\n","    return p\n","\n","#Loading the viewer by indicating the MSA file and format to read\n","#@markdown Name of the MSA file (including the filetype)\n","MSAfile = 'aligned.fasta' #@param {type:\"string\"}\n","MSAformat = 'fasta' #@param {type:\"string\"}\n","aln = AlignIO.read(MSAfile,MSAformat)\n","p = view_alignment(aln, plot_width=900)\n","pn.pane.Bokeh(p)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"exUsc7YR-dXe"},"source":["3. As you can see, the sequences are aligned, as many residues that are conserved between different sequences occupy the same columns within the alignment. However, **we can also see\n","some sequences that are longer on the ends of the protein**. These ends will contribute nothing to the alignment, as there is nothing to compare them to. Therefore, we will trim the sequences on both N-and C-ends by selecting the regions we want to eliminate.\n","\n","  Ideally, you would manually correct the alignment and trim the ends of the sequences, and **we encourage you to do it**. Due to time restrictions, we include a script below that trims the N- and C-ends of the alignment based on finding the first and lasts columns of the MSA without any gaps.\n"]},{"cell_type":"code","metadata":{"id":"cPgR1wuGzHWL"},"source":["import sys\n","from Bio import AlignIO\n","aln = AlignIO.read(\"aligned.fasta\", \"fasta\")\n","\n","for fcol in range(aln.get_alignment_length()):\n","  if not \"-\" in aln[:, fcol]:\n","    position1 = fcol\n","    print(\"First full column is {}\".format(fcol))\n","    break\n","for lcol in reversed(range(aln.get_alignment_length())):\n","  if not \"-\" in aln[:, lcol]:\n","    position2 = lcol+1\n","    print(\"Last full column is {}\".format(lcol))\n","    break\n","\n","print(\"New alignment:\")\n","print(aln[:, position1:position2])\n","\n","with open(\"aligned_trimmed.fasta\", \"w\") as handle:\n","  count = (SeqIO.write(aln[:, position1:position2], handle, \"fasta\"))\n","\n","trim = AlignIO.read(\"aligned_trimmed.fasta\", \"fasta\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"uwIFOU8L_rAg"},"source":["4. We also need to **make sure that there are no duplicated sequences included in it**. Having redundant information does not have any benefit for the phylogenetic analysis that we are pursuing here.\n","\n","  Do we have duplicated sequences? We can rapidly generate a distance matrix based on the pairwise differences between all sequences in the alignment file using the biopython _DistanceCalculator_ tool. Thus, a pair of sequences with 0.0 differences would correspond to equivalent sequences (100% sequence identity)."]},{"cell_type":"code","metadata":{"id":"xm7b8rCAAC8O"},"source":["from Bio.Phylo.TreeConstruction import DistanceCalculator\n","from Bio import AlignIO\n","aln = AlignIO.read(open('aligned_trimmed.fasta'), 'fasta')\n","calculator = DistanceCalculator('identity')\n","dm = calculator.get_distance(aln)\n","print(dm)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"9x1cNdED0rgb"},"source":["5. There are suitable programs to eliminate the duplicated sequences that you detected in the exercise above, such as **CD-HIT**, but here we opted to directly execute another script that compares the sequences in the alignment and eliminates the duplicated ones."]},{"cell_type":"code","metadata":{"id":"XcPGJREX0057"},"source":["#Sequence cleaner script\n","#Modified from https://peterjc.github.io/wiki/Sequence_Cleaner\n","from Bio import SeqIO\n","\n","def sequence_cleaner(fasta_file, min_length=0):\n","  # Create our hash table to add the sequences\n","  sequences = {}\n","  # Using the Biopython fasta parse we can read our fasta input\n","  for seq_record in SeqIO.parse(fasta_file, \"fasta\"):\n","    # Take the current sequence\n","    sequence = str(seq_record.seq).upper()\n","    # Check if the current sequence is according to the user parameters\n","    if (len(sequence) >= min_length):\n","      # If the sequence passed in the test \"is it clean?\" and it isn't in the\n","      # hash table, the sequence and its id are going to be in the hash\n","        if sequence not in sequences:\n","          sequences[sequence] = seq_record.id\n","      # If it is already in the hash table, we're just gonna concatenate the ID\n","      # of the current sequence to another one that is already in the hash table\n","        else:\n","          sequences[sequence] += \"_\" + seq_record.id\n","\n","  # Write the clean sequences\n","  # Create a file in the same directory where you ran this script\n","  with open(\"clear_\" + fasta_file, \"w+\") as output_file:\n","  # Just read the hash table and write on the file as a fasta format\n","    for sequence in sequences:\n","      output_file.write(\">\" + sequences[sequence] + \"\\n\" + sequence + \"\\n\")\n","  print(\"CLEAN!!!\\nPlease check clear_\" + fasta_file)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"fQYoHOUK1BnK"},"source":["sequence_cleaner('aligned_trimmed.fasta', 0)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"lSIIvNPzAktp"},"source":["**QUESTION:** What would happen with our phylogenetic analysis and phylogenetic tree if we did not delete these duplicated sequences?\n","\n","6. Our aligned, trimmed and curated MSA looks as shown below:"]},{"cell_type":"code","metadata":{"cellView":"form","id":"KGnHXeHAgQi_"},"source":["#@title Protein MSA Viewer in Google Colab\n","#The following code is modified from the wonderful viewer developed by Damien Farrell\n","#https://dmnfarrell.github.io/bioinformatics/bokeh-sequence-aligner\n","\n","#Importing all modules first\n","import os, io, random\n","import string\n","import numpy as np\n","\n","from Bio.Seq import Seq\n","from Bio.Align import MultipleSeqAlignment\n","from Bio import AlignIO, SeqIO\n","\n","import panel as pn\n","import panel.widgets as pnw\n","pn.extension()\n","\n","from bokeh.plotting import figure\n","from bokeh.models import ColumnDataSource, Plot, Grid, Range1d\n","from bokeh.models.glyphs import Text, Rect\n","from bokeh.layouts import gridplot\n","\n","#Setting up the amino color code according to Zappo color scheme\n","def get_colors(seqs):\n","    #make colors for bases in sequence\n","    text = [i for s in list(seqs) for i in s]\n","    #Use Zappo color scheme\n","    clrs =  {'K':'red',\n","             'R':'red',\n","             'H':'red',\n","             'D':'green',\n","             'E':'green',\n","             'Q':'blue',\n","             'N':'blue',\n","             'S':'blue',\n","             'T':'blue',\n","             'A':'blue',\n","             'I':'blue',\n","             'L':'blue',\n","             'M':'blue',\n","             'V':'blue',\n","             'F':'orange',\n","             'Y':'orange',\n","             'W':'orange',\n","             'C':'blue',\n","             'P':'yellow',\n","             'G':'orange',\n","             '-':'white'}\n","    colors = [clrs[i] for i in text]\n","    return colors\n","\n","#Setting up the MSA viewer\n","def view_alignment(aln, fontsize=\"9pt\", plot_width=800):\n","    \"\"\"Bokeh sequence alignment view\"\"\"\n","\n","    #make sequence and id lists from the aln object\n","    seqs = [rec.seq for rec in (aln)]\n","    ids = [rec.id for rec in aln]\n","    text = [i for s in list(seqs) for i in s]\n","    colors = get_colors(seqs)\n","    N = len(seqs[0])\n","    S = len(seqs)\n","    width = .4\n","\n","    x = np.arange(1,N+1)\n","    y = np.arange(0,S,1)\n","    #creates a 2D grid of coords from the 1D arrays\n","    xx, yy = np.meshgrid(x, y)\n","    #flattens the arrays\n","    gx = xx.ravel()\n","    gy = yy.flatten()\n","    #use recty for rect coords with an offset\n","    recty = gy+.5\n","    h= 1/S\n","    #now we can create the ColumnDataSource with all the arrays\n","    source = ColumnDataSource(dict(x=gx, y=gy, recty=recty, text=text, colors=colors))\n","    plot_height = len(seqs)*15+50\n","    x_range = Range1d(0,N+1, bounds='auto')\n","    if N>100:\n","        viewlen=100\n","    else:\n","        viewlen=N\n","    #view_range is for the close up view\n","    view_range = (0,viewlen)\n","    tools=\"xpan, xwheel_zoom, reset, save\"\n","\n","    #entire sequence view (no text, with zoom)\n","    p = figure(title=None, width= plot_width, height=50,\n","               x_range=x_range, y_range=(0,S), tools=tools,\n","               min_border=0, toolbar_location='below')\n","    rects = Rect(x=\"x\", y=\"recty\",  width=1, height=1, fill_color=\"colors\",\n","                 line_color=None, fill_alpha=0.6)\n","    p.add_glyph(source, rects)\n","    p.yaxis.visible = False\n","    p.grid.visible = False\n","\n","    #sequence text view with ability to scroll along x axis\n","    p1 = figure(title=None, width=plot_width, height=plot_height,\n","                x_range=view_range, y_range=ids, tools=\"xpan,reset\",\n","                min_border=0, toolbar_location='below')#, lod_factor=1)\n","    glyph = Text(x=\"x\", y=\"y\", text=\"text\", text_align='center',text_color=\"black\",\n","                text_font=\"monospace\",text_font_size=fontsize)\n","    rects = Rect(x=\"x\", y=\"recty\",  width=1, height=1, fill_color=\"colors\",\n","                line_color=None, fill_alpha=0.4)\n","    p1.add_glyph(source, glyph)\n","    p1.add_glyph(source, rects)\n","\n","    p1.grid.visible = False\n","    p1.xaxis.major_label_text_font_style = \"bold\"\n","    p1.yaxis.minor_tick_line_width = 0\n","    p1.yaxis.major_tick_line_width = 0\n","\n","    p = gridplot([[p],[p1]], toolbar_location='below')\n","    return p\n","\n","#Loading the viewer by indicating the MSA file and format to read\n","#@markdown Name of the MSA file (including the filetype)\n","MSAfile = 'clear_aligned_trimmed.fasta' #@param {type:\"string\"}\n","MSAformat = 'fasta' #@param {type:\"string\"}\n","aln = AlignIO.read(MSAfile,MSAformat)\n","p = view_alignment(aln, plot_width=900)\n","pn.pane.Bokeh(p)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"4JH1AQWch94m"},"source":["#Part III - Phylogenetic inference and ancestral sequence reconstruction using RAxML"]},{"cell_type":"markdown","metadata":{"id":"9AX2DwfeBs0y"},"source":["We will use our final, trimmed and filtered alignment to infer the phylogenetic relationships of our sequences with the **heuristic** tree-searching method of **RAxML** using the **Maximum Likelihood (ML)** optimality criterion.\n","\n","While we already discussed about some of the phylogenetic methods as well as evolutionary models during our Lecture, **we invite you again to read the Lecture Appendix on ML.**\n","\n","What we did not discuss was that the evolutionary models could also incorporate additional parameters, such as invariant sites (conservation) or variations in the substitution rates across sites. For example, **initiation codons (typically, ATG)** may not be free to vary at all. Therefore, we should make such sites invariant. On the other hand, we know that positions on the interior of a protein evolve more slowly that surface residues, therefore substitutions have **varying rates** depending on the position of these residues. In the case of variation, modeling these varying rates is computationally consuming, thus a well-behaved, mathematically tractable **gamma distribution** is used for modeling these rate variations."]},{"cell_type":"markdown","metadata":{"id":"ohAIWzroCvGg"},"source":["1. Now we will starting doing some ML phylogenetics, but **which model should we use?** Well, ML has the advantage that we can use the **log-likelihood** to infer which evolutionary model is more suitable for the alignment being used (neat, isn’t it?). This is implemented in the **ModelTest-NG** program and can be used as follows:\n"]},{"cell_type":"code","metadata":{"id":"klKC-g6dCKqy"},"source":["!modeltest-ng -i clear_aligned_trimmed.fasta -d aa"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"FRmZhhB67_ep"},"source":["The result will look very complex and unreadable for you now. However, the models are selected according to **AICc**, which determines the suitability of a given model if it best-minimize **-lnL** with the smallest number of parameters. Here, **lnL** is the log-likelihood, with its negative value used as the minimization target during model selection. **In our particular case, the best model turned out as LG+I+G4**, or the **Le-Gascuel model** with **Gamma** variation and **a proportion of invariable sites**. Both gamma variation and invariable sites add rate heterogeneity across sites."]},{"cell_type":"markdown","metadata":{"id":"w1eiPJIF90dU"},"source":["2. Now, we can use our final MSA and this evolutionary model to perform our first phylogenetic analysis using **RAxML-NG**"]},{"cell_type":"code","metadata":{"id":"FHKS7xYHEVh_"},"source":["!raxml-ng --msa clear_aligned_trimmed.fasta --model LG+I+G4 --prefix T1 --threads 2"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"tS0DiOPc-jg7"},"source":["3. We can again use biopython to see the results of our phylogenetic inference using the _Phylo_ tool."]},{"cell_type":"code","metadata":{"id":"f-elKCfF-s8n"},"source":["from Bio import Phylo\n","import matplotlib.pyplot as plt\n","\n","# Read the tree\n","tree = Phylo.read(\"T1.raxml.bestTree\", \"newick\")\n","\n","# Set up the figure size manually with a new figure\n","fig, ax = plt.subplots(figsize=(10, 6))  # Adjust the size as needed\n","\n","# Remove the top and right spines (axes lines)\n","ax.spines['right'].set_visible(False)\n","ax.spines['top'].set_visible(False)\n","\n","# Draw the tree on the axes with the size specified above\n","Phylo.draw(tree, axes=ax)\n","\n","# Display the figure\n","plt.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Eecsg5R_-1GS"},"source":["**QUESTION:** How many clades do you see? How many common ancestors there are? Who are the members of the clade containing our query sequence?"]},{"cell_type":"markdown","metadata":{"id":"llzvGVk8EadP"},"source":["4. Now we need to test the reliability (reproducibility) of our phylogenetic tree using the **boostrapping method** that we discussed during our Lectures. Remember that this is not part of the tree construction method, but a phylogeny test.\n","\n","  While usually 1000-2000 bootstrap replicates are suggested for determining the confidence of the phylogenetic tree, RAxML also has a so-called **bootstopping** method that determines how many bootstrap replicates are required to obtain stable support values. However, for a speedy tutorial, we have requested 100 boostrap replicates only."]},{"cell_type":"code","metadata":{"id":"Ov41LC47BUNe"},"source":["!raxml-ng --bootstrap --msa T1.raxml.rba --model LG+I+G4 --prefix T2 --threads 2 --bs-tree 100"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"jKPkOtxjE2ap"},"source":["**QUESTION:** A reliable clade should have bootstrapping values **> 70**. What would a bootstrapping value of 70 mean?\n","\n","**💡 HINT:** You can still check the convergence of the boostrapping test afterwards by executing the following command:\n","\n","```\n","!raxml-ng --bsconverge --bs-trees T2.raxml.bootstraps --prefix Test --threads 2 --bs-cutoff 0.03\n","```\n","\n","In practice, a convergence cutoff value of 3% should be sufficient in most cases.\n"]},{"cell_type":"markdown","metadata":{"id":"IMWzFYCJCnaR"},"source":["5. Now, we will map the support values obtained from the boostrapping test onto the best-scoring ML tree on the original MSA. Once you ran the cell code below, use the Phylo package again to show the phylogenetic tree with its bootstrapping values."]},{"cell_type":"code","metadata":{"id":"3Rr3dZmWHDVV"},"source":["!raxml-ng --support --tree T1.raxml.bestTree --bs-trees T2.raxml.bootstraps --prefix T3 --threads 2"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"9uwp7HC7ClwT"},"source":["#Here, use the biopython Phylo package again!\n","from Bio import Phylo\n","import matplotlib.pyplot as plt\n","tree = Phylo.read(\"T3.raxml.support\", \"newick\")\n","fig, ax = plt.subplots(figsize=(10, 6))  # Adjust the size as needed\n","ax.spines['right'].set_visible(False)\n","ax.spines['top'].set_visible(False)\n","Phylo.draw(tree, axes=ax)\n","plt.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"j6Q-WDyIH8K9"},"source":["6. With this, we are done with inferring our phylogenetic trees. Also, we can get and download the **YourFileName.raxml.bestTree** file (which is the tree file) an see it in the visualizing website [**iTOL**](https://itol.embl.de/).\n","\n","  At the top of this site, there is an **UPLOAD** button. We can press the button and upload the besttree file to nicely draw our phylogenetic tree."]},{"cell_type":"markdown","metadata":{"id":"Lnl-9luRIgY2"},"source":["7. We will now **use the ML tree** to obtain an estimation of the **ancestral sequence** for the FoxP clade. The DNA encoding these protein sequences are usually synthesized by researchers to evaluate their structure and function and allowing inference of the cellular, environmental and functional contexts of ancestral and extant organisms.\n","\n","  Since we already have the MSA and the ML tree, we do not need anything extra to obtain the ancestral sequences at each of the internal nodes. Just use the code below.\n"]},{"cell_type":"code","metadata":{"id":"GmxWkKnJiLaO"},"source":["!raxml-ng --ancestral --msa clear_aligned_trimmed.fasta --tree T1.raxml.bestTree --model LG+I+G4 --prefix ASR"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"7cgzhP-dJN6b"},"source":["8. Now, open the **YourFileName.raxml.ancestralStates** to see the most probable sequences based on the probabilities for each position of the alignment (which are contained in the **YourFileName.raxml.ancestralProbs** file).\n","\n","  But which sequence correspond to which node? You can now draw the **YourFileName.raxml.ancestralTree** to check the numbers of each node and determine which sequence correspond to the FoxP clade."]},{"cell_type":"code","metadata":{"id":"zBvCa7KGKGEL"},"source":["#Here, use the biopython Phylo package again!\n","from Bio import Phylo\n","import matplotlib.pyplot as plt\n","tree = Phylo.read(\"ASR.raxml.ancestralTree\", \"newick\")\n","fig, ax = plt.subplots(figsize=(10, 6))  # Adjust the size as needed\n","ax.spines['right'].set_visible(False)\n","ax.spines['top'].set_visible(False)\n","Phylo.draw(tree, axes=ax)\n","plt.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"jJj1pCHaK7BS"},"source":["**📚 HOMEWORK:** Now, as a final exercise, please compare the sequence of the ancestral node for the FoxP clade with the sequences shown in the Overview for FoxP and for the FoxP homologs. Also, take a look at the structures of the different Fox homologs that you have in your alignment by searching for the accession codes in the [Protein Data Bank](https://www.rcsb.org).\n","\n","  Based on the information provided in the Overview of this tutorial, the phylogenetic analysis of FoxP proteins and their homologs, the reconstructed ancestral sequences for each node and your observations of the different structures that are included in your sequence alignment:\n","\n","1. Can we say something about the most probable oligomeric state of this protein based on this sequence?\n","\n","2. Can we say something about how this subfamily gained the ability to dimerize during the evolution of Fox proteins?\n","\n","3. Is there any element that we did not take into account that could significantly alter the results from our analysis?\n","\n","\n","**This is the end of the third tutorial. Good science!**"]}]}
