{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"TTn0im17RKAk"},"source":["#Lab.12 / IBM3202 – Protein folding *ab initio* using Rosetta"]},{"cell_type":"markdown","metadata":{"id":"HImm2gBRR6Xf"},"source":["#Theoretical aspects.\n","Why can we perform structure predictions for a given protein sequence?"]},{"cell_type":"markdown","metadata":{"id":"ARYtS662SC-T"},"source":["As we already saw in the homology modelling tutorial there is a growing need to solve the 3D structure of proteins. As evolution has evolved from a finite number of families of proteins there are usually closely related proteins that has been already solved and can be used in strategies as the aforementioned homology modelling. But there is also a chance of you encountering a protein with no closely related homologs which has been already structurally solved, in this last case scenario we can use modelling derived from first principles or more commonly refered as *Ab initio* modelling, which is template agnostic.\n","\n"]},{"cell_type":"markdown","metadata":{"id":"_HRMjMeeTfOs"},"source":["As we have discussed during the course, there are many ways you can predict a protein’s structure by knowing only its sequence: homology-based comparative modelling, determination of coevolutionary couplings using a myriad of homologous protein sequences and ab initio structure prediction. Ab initio comes from the Latin and means “from first principles”, in which the idea is to arrive at a protein structure just by knowing the principles governing protein folding."]},{"cell_type":"markdown","metadata":{"id":"IKjxOMBoaqqc"},"source":["Where do these principles come from? As we have seen in our lectures, the ever-increasing number of solved protein structures by different methods (X-ray, NMR, cryoEM) has enabled to define that, regardless their primary sequence, proteins do not explore all possible configurations. This very simple observation enabled further use of this information to perform statistical analysis of the different conformations that a protein sequences usually explore: typical amino acid environments and distances between amino acids in any given protein, secondary structure predictions based on sequence information alone, angle and distance geometries between secondary structure elements, dihedral distributions of backbones and side chains, etc. Then, this information can be used to define energy functions which enable us to determine which structures are more probable for a given sequence (an inverse folding problem).\n"]},{"cell_type":"markdown","metadata":{"id":"ojBSonugasAx"},"source":["While this method has been used for several protein structure predictions, the exponential growth of the conformational space of a protein that can be explored by a given protein sequence with respect to its protein length severely limits its use to proteins with fewer than 100 residues. The rationale behind such exact number is that a protein domain, i.e. considered as a region (sequence) of a protein that can fold independently onto a given structure by its own, typically has a maximum in its typical length distribution at ~100 residues for most proteins (**Fig 1**). Thus, there seems to be an optimal number of residues that allow proteins to form a hydrophobic core big enough to stably fold."]},{"cell_type":"markdown","metadata":{"id":"HBp5K9IOsQzi"},"source":["<figure>\n","<center>\n","<img src=\"https://ars.els-cdn.com/content/image/1-s2.0-S1359027898000042-gr3.gif\"/>\n","\n","<figcaption>FIGURE 1. Length distribution of proteins domains. The arrow represents the maximum of the distribution, which in turn represents the minimum of a folding energy function that depends solely on the entropic cost of packing a n-polymer, the hydration cost for its radios of gyration, and the average internal enthalpy of contact formation.  <br> Adapted from Xu & Nussinov (1998)<i> Folding & Design.</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"RoM2js2obWYH"},"source":["Although this is used as a general rule, there are plenty of exceptions as shown in the above distribution, of domains that form well below or beyond the optimal 100 ± 40 residues. Some extremely short proteins can adopt stable structures in solution (i.e. the β-stable hairpin we studied through MD, or the villin head we are about to fold), and immensely large proteins that can form a huge hydrophobic core. This noisy distribution highlights that there is not a unique solution to the protein folding problem, and the principles governing the folding of ideal, regular-sized globular proteins may not be the same than that of outlier, small domains. Today, we will put that to the test."]},{"cell_type":"markdown","metadata":{"id":"W8Q2C-sBblS4"},"source":["#General Overview"]},{"cell_type":"markdown","metadata":{"id":"VNpLsgP1boQf"},"source":["In this tutorial, we will employ Rosetta to predict the protein structure of the 35-residue headpiece of villin from G. gallus. Although compiling may be messy, and the documentation is awfully dispersed and obscure through tens of different versions and small unannounced modifications, Rosetta is by far the most successful engine in de novo design and prediction of proteins structures (Kaufmann *et al.* (2010) *Biochemistry*).\n","\n","Briefly, we will perform the following tasks:\n","\n","Part I – Prepare the fragment libraries required for protein structure prediction using Rosetta\n","\n","Part II – Execute Rosetta to generate 10 ab initio models\n","\n","Part III – Analyze the results using py3Dmol and Rosetta\n","\n","Part IV – Prepare a larger number of models (>10,000) using two different fragment libraries"]},{"cell_type":"markdown","metadata":{"id":"Q55MC985bhoJ"},"source":["##Folding protocol overview\n"]},{"cell_type":"markdown","metadata":{"id":"K4r7VDTkULy0"},"source":["As stated in the **AbinitioRelax [documentation](https://)**\n","\n","> The AbinitioRelax application consists of two main steps.\n","1. The first step is a coarse-grained fragment-based search through conformational space using a knowledge-based \"centroid\" score function that favors protein-like features (Abinitio).\n","2. The second optional step is all-atom refinement using the Rosetta full-atom forcefield (Relax). The \"Relax\" step is considerably more compute-intensive and time-consuming than the first step.\n","\n","There is also commented the fact that:\n","\n","> For increased conformational sampling, this application is easily parallelized by executing numerous jobs each using a unique random number seed (see Options section below). This is typically done by submitting multiple jobs to a computer cluster or distributed grid. Since the full-atom energy function is very sensitive to imperfect atomic interactions and more noise will exist with insufficient sampling, convergence towards the native structure may require a significant amount of sampling. Additionally, to increase your chance of sampling the correct topology, a diverse set of homologous sequences, preferably with sequence changes that may have a greater impact on sampling like deletions and differences in conserved positions, may also be run since a homologue may converge towards the native structure with significantly less sampling (See Bradley *et al.* (2005). Toward high-resolution de novo structure prediction for small proteins. *Science*, 309(5742), 1868-1871).\n","\n","\n","\n","\n"]},{"cell_type":"markdown","metadata":{"id":"fQbd_TeklIb0"},"source":["## Folding.py\n","This script performs fragment insertion for an input protein sequence. The\n","sequence may be explicit, in a FASTA file, or in a PDB file. Two fragment files,\n","preferably one longer than the other, must also be provided. Fragment insertion\n","is accompanied by a Monte Carlo assessment allowing the conformation to escape\n","local minima. Output structures from this protocol are intended to proceed into\n","a refinement step (such as that in refinement.py) to produce reasonable\n","estimates of the protein conformation.\n","\n","Instructions:\n","\n","1. ensure that your PDB file is in the current directory\n","2. run the script:\n","  - from commandline  \n","      - > python D060_Folding.py\n","  - from within python/ipython              \n","      - run D060_Folding.py\n","\n","Author: Evan H. Baugh\n","    revised and motivated by Robert Schleif\n","\n","Updated by Boon Uranukul, 6/9/12\n","Simplified special constant seed initialization ~ Labonte\n","\n","References:\n","    P. Bradley, K. Misura, and D. Baker, \"Toward high-resolution de novo\n","        structure prediction for small proteins\", *Science* 309 (5742)\n","        1868-1871 (2005).\n","\n","\"\"\""]},{"cell_type":"markdown","metadata":{"id":"sCahXpkxlX-T"},"source":["## The complete folding protocol, steps included in the \"sample_folding\" definition"]},{"cell_type":"markdown","metadata":{"id":"vqzFPKS9lVLg"},"source":["\n","\n","This sample script is setup for usage with\n","    commandline arguments,\n","    default running within a python interpreter,\n","    or for import within a python interpreter,\n","        (exposing the sample_folding method)\n","\n","The method sample_folding:\n","1.  creates a pose from the desired sequence (fullatom)\n","2.  set all of the pose's backbone torsion angles\n","3.  creates a (fullatom) reference copy of the pose\n","4.  creates Movers for switching between fullatom and centroid\n","5.  convert both poses to centroid\n","6.  creates a MoveMap with all backbone torsion angles free\n","7.  sets up ClassicFragmentMovers for inserting fragments backbone structures for long and short fragments\n","8.  creates RepeatMovers for the long and short fragment Movers\n","9.  create a PyMOL_Observer for viewing intermediate output\n","10. creates low and high resolution ScoreFunctions\n","11. sets up a RepeatMover on a TrialMover of a SequenceMover setup the TrialMover\n","      \n","      -  create a SequenceMover with the:\n","          - RepeatMover on the long fragment Mover\n","          - RepeatMoversepeatMover on the short fragment Mover\n","\n","      -  create a MonteCarlo object for assessing moves\n","      -  create the TrialMover (on the SequenceMover)\n","      -  create the RepeatMover (on the TrialMover)\n","12. creates a (Py)JobDistributor for managing multiple trajectories\n","13. stores the original score evaluation (centroid)\n","14. performs low-resolution folding:\n","  - set necessary variables\n","      - reload the starting (centroid) pose\n","      - change the pose's PDBInfo.name, for the PyMOLMover\n","      - reset the MonteCarlo object (for the TrialMover)\n","  - perform sampling and assessment using the final RepeatMover\n","  - convert the lowest scoring decoy to fullatom using the SequenceMover\n","  - export the (lowest scoring) decoy structure\n","      - recover the lowest scoring (centroid) decoy\n","      - store the decoy score\n","      - convert the decoy to fullatom\n","      - guess potential disulfide bridges\n","      - output the decoy structure using the PyJobDistributor\n","      - export the decoy structure using the PyMOL_Observer\n","15. outputs the score evaluations\n","\n","The second method, guess_disulfides, is called in sample_folding to print out\n","cysteine residues which are close to each other in the decoy structure.\n"]},{"cell_type":"markdown","metadata":{"id":"yZeHMrpWcdzz"},"source":["#Part I – Prepare the fragment libraries required for protein structure prediction using Rosetta"]},{"cell_type":"markdown","metadata":{"id":"gO0B8VE3cj-n"},"source":["The sequence of the protein for which we want to predict its structure, the **G34L mutant** of the headpiece subdomain of the **Villin** protein from *G. gallus*, is the following:\n","> MLSDEDFKAFGMTRSAFANLPLWKQQNLKKEKLLF\n","\n","The first thing we need to do is to generate a fragment library that we will require for modelling our protein. As you remember from the lectures, these libraries are essential for closing loops between segments and for determining the most probable local structure for a given sequence.\n","\n","To generate these libraries, we recommend using **Robetta (http://old.robetta.org/)**, a webserver that is kept by the same laboratory in charge of maintaining Rosetta. Here, one typically will go onto the “Submit” section of the Fragment Library service, and then input the name of the target protein in the “Target Name” field and its full sequence in Fasta format in the “Paste Fasta” field. Finally, we would normally press “Submit” to start the fragment library generation."]},{"cell_type":"markdown","metadata":{"id":"gRpdqFS7c0xK"},"source":["Depending on the protein sequence length, these fragment library generations can take up to several hours. Therefore, we are providing these files for your uninterrupted work (0- fragment_libraries/villin folder). If you take a quick look onto these files, you should see:\n","\n","* 3-residue long fragment file (i.e. aat000_03...)\n","\n","* 9-residue long fragment file (i.e. aat000_09...)\n","\n","* 3 temporal files (i.e. *.dat, *.check, *.checkpoint)\n","\n","* a Fasta file with your sequence (i.e. *.fasta)\n","\n","* 3 secondary structure prediction files (i.e. *.jufo_ss, *.psipred, *.psipred_ss2)\n","\n","* a database file (i.e. *.rdb)"]},{"cell_type":"markdown","metadata":{"id":"NyWAWqFxddgs"},"source":["We are only interested in the fragment files. Open one of them in a text editor to examine its content. For each, 200 fragments will be generated, and the content is as follows:\n"]},{"cell_type":"markdown","metadata":{"id":"8VFxin_LdgVd"},"source":["* Column -- Meaning\n","* 1 -- blank\n","\n","* 2-5 -- PDB code for the fragment origin\n","\n","* 7 -- chain ID for the origin PDB\n","\n","* 9-13 -- PDB residue number for the origin PDB\n","\n","* 15 -- amino acid identity in the origin PDB\n","\n","* 17 -- secondary structure for the origin PDB (Helix, Loop, Extended/beta) 19-27 -- phi\n","\n","* 28-36 -- psi\n","\n","* 37-45 -- omega\n","\n","* 46-54 -- C-alpha x coordinate for origin PDB (optional)\n","\n","* 55-63 -- C-alpha y coordinate for origin PDB (optional)\n","\n","* 65-73 -- C-alpha z coordinate for origin PDB (optional)\n","\n","* 74-79 -- unknown (unused)\n","\n","* 80-85 -- unknown (unused)\n","\n","* 86 -- Literal \"P\" (unused)\n","\n","* 87-89 -- fragment position number, pose numbered (unused)\n","\n","* 91 -- Literal \"F\"(unused)\n","\n","* 92-94 -- fragment number (unused)\n","\n","**QUESTION!!**\n","* Why are the position coordinate of the fragments optional?\n","* Why we do not require any information about the structural coordinates of the rest of the atoms?"]},{"cell_type":"markdown","metadata":{"id":"QVHm1u1pmrvp"},"source":["# Part II – Execute Rosetta to generate 10 *ab initio* models\n","\n","Once we have generated the fragment libraries, we will employ them along with Rosetta. In its core, Rosetta is quite similar to Gromacs in that it consists of a series of individual applications or codes that share in common a language and libraries. In this case, we will use the application named **AbinitioRelax**. To examine this application in detail, please Google the name of the application followed by Rosetta, and you’ll immediately find the documentation page alongside with useful references. **Do it. Seriously. Rosetta is an extremely convoluted and complicated program!**"]},{"cell_type":"code","metadata":{"id":"BasFTps1bPtS"},"source":["# Notebook setup\n","import sys\n","!pip install pyrosettacolabsetup\n","import pyrosettacolabsetup; pyrosettacolabsetup.install_pyrosetta()\n","import pyrosetta; pyrosetta.init()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"VAaC0ezm5lHD"},"source":["Here we copy test files for running the method"]},{"cell_type":"code","metadata":{"id":"9DrGx1TfeQUQ"},"source":["#Here we download the input files from Github\n","!wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/input_files/input_files_ai.zip\n","#Here, we unzip the downloaded files\n","!unzip input_files_ai.zip"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"5RH87G8R7JS5"},"source":["Here we create the outs folder needed for saving the results from folding"]},{"cell_type":"code","metadata":{"id":"2NQMSFmM1eAa"},"source":["!mkdir outs"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ubELeWHpqxEi"},"source":["Following is the protocol for running from colab"]},{"cell_type":"markdown","metadata":{"id":"9sduT7I50orJ"},"source":["### Definition of method for folding protocol"]},{"cell_type":"code","metadata":{"id":"QiU7m58HOCjQ"},"source":["import optparse    # for sorting options\n","\n","import pyrosetta\n","import pyrosetta.rosetta as rosetta\n","\n","from pyrosetta import (\n","    init, pose_from_sequence, pose_from_file, Pose, MoveMap, create_score_function, get_fa_scorefxn,\n","    MonteCarlo, TrialMover, SwitchResidueTypeSetMover, PyJobDistributor,\n",")\n","from pyrosetta.rosetta import core, protocols\n","init(extra_options = \"-scorefile_format text\")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"XxL-i1AhqsOW"},"source":["def sample_folding(sequence,\n","        long_frag_filename, long_frag_length,\n","        short_frag_filename, short_frag_length,\n","        kT = 3.0, long_inserts = 1, short_inserts = 3, cycles = 40,\n","        jobs = 1, job_output = 'fold_output'):\n","    \"\"\"\n","\n","    \"\"\"\n","    # 1. create a pose from the desired sequence (fullatom)\n","    # the method pose_from_sequence produces a complete IDEALIZED\n","    #    protein conformation of the input sequence, the ResidueTypeSet (second\n","    #    argument below) may be varied, and this method supports non-proteogenic\n","    #    chemistry (though it is still a Rosetta Residue). however this syntax\n","    #    is more involved and not robust to user errors, and not presented here\n","    # small differences in bond lengths and bond angles WILL change the results,\n","    #### if you desire an alternate starting conformation, alter steps\n","    ####     1. and 2. as you please\n","    pose = pose_from_sequence(sequence, 'fa_standard')\n","\n","    # 2. linearize the pose by setting backbone torsions to large values\n","    # the method make_pose_from_sequence does not create the new pose's\n","    #    PDBInfo object, so its done here, without it an error occurs later\n","    pose.pdb_info(rosetta.core.pose.PDBInfo( pose.total_residue() ))\n","    for i in range(1, pose.total_residue() + 1):\n","        pose.set_omega(i, 180)\n","        pose.set_phi(i, -150)    # reasonably straight\n","        pose.set_psi(i, 150)\n","        #### if you want to see the decoy scores, the PDBInfo needs these lines\n","        #pose.pdb_info().chain(i, 'A')    # necessary to color by score\n","        #pose.pdb_info().number(i, i)    # for PDB numbering\n","    ####\n","\n","    # 3. create a (fullatom) reference copy of the pose\n","    test_pose = Pose()\n","    test_pose.assign( pose )\n","    test_pose.pdb_info().name('linearized pose')\n","\n","    # 4. create centroid <--> fullatom conversion Movers\n","    to_centroid = SwitchResidueTypeSetMover('centroid')\n","    # centroid Residue objects, of amino acids, have all their sidechain atoms\n","    #    replaced by a single representative \"atom\" to speed up calculations\n","    to_fullatom = SwitchResidueTypeSetMover('fa_standard')\n","\n","    # 5. convert the poses to centroid\n","    to_centroid.apply(pose)\n","    to_centroid.apply(test_pose)\n","\n","    # 6. create the MoveMap, all backbone torsions free\n","    movemap = MoveMap()\n","    movemap.set_bb(True)\n","    # minimizing the centroid chi angles (the sidechain centroid atoms) is\n","    #    almost always USELESS since this compression is performed for speed,\n","    #    not accuracy and clashes usually occur when converting to fullatom\n","\n","    # 7. setup the ClassicFragmentMovers\n","    # for the long fragments file\n","    # this \"try--except\" is used to catch improper fragment files\n","    try:\n","        fragset_long = core.fragment.ConstantLengthFragSet( long_frag_length , long_frag_filename )\n","        #### the ConstantLengthFragSet is overloaded, this same\n","        ####    ConstantLengthFragSet can be obtained with different syntax\n","        # to obtain custom fragments, see Generating Fragment Files below\n","    except:\n","        raise IOError('Make sure long_frag_length matches the fragments in\\n\\\n","            long_frag_file and that long_frag_file is valid')\n","    long_frag_mover = protocols.simple_moves.ClassicFragmentMover(fragset_long, movemap)\n","    # and for the short fragments file\n","    # this \"try--except\" is used to catch improper fragment files\n","    try:\n","        fragset_short = core.fragment.ConstantLengthFragSet(short_frag_length, short_frag_filename)\n","    except:\n","        raise IOError('Make sure short_frag_length matches the fragments in\\n\\\n","            short_frag_file and that short_frag_file is valid')\n","    short_frag_mover = protocols.simple_moves.ClassicFragmentMover(fragset_short, movemap)\n","\n","    # 8. setup RepeatMovers for the ClassicFragmentMovers\n","    insert_long_frag = protocols.moves.RepeatMover(long_frag_mover, long_inserts)\n","    insert_short_frag = protocols.moves.RepeatMover(short_frag_mover, short_inserts)\n","\n","    # 9. create a PyMOL_Observer for exporting structures to PyMOL (optional)\n","    # the PyMOL_Observer object owns a PyMOLMover and monitors pose objects for\n","    #    structural changes, when changes are detected the new structure is\n","    #    sent to PyMOL\n","    # fortunately, this allows investigation of full protocols since\n","    #    intermediate changes are displayed, it also eliminates the need to\n","    #    manually apply the PyMOLMover during a custom protocol\n","    # unfortunately, this can make the output difficult to interpret (since you\n","    #    aren't explicitly telling it when to export) and can significantly slow\n","    #    down protocols since many structures are output (PyMOL can also slow\n","    #    down if too many structures are provided and a fast machine may\n","    #    generate structures too quickly for PyMOL to read, the\n","    #    \"Buffer clean up\" message\n","    # uncomment the line below to use PyMOL_Observer\n","    ##    AddPyMOLObserver(test_pose, True)\n","\n","    # 10. create ScoreFunctions\n","    # for low-resolution, centroid, poses necessary for the TrialMover's\n","    #    MonteCarlo object (see below)\n","    scorefxn_low = create_score_function('score3')\n","    # for high-resolution, fullatom, poses necessary for scoring final output\n","    #    from the PyJobDistributor (see below)\n","    scorefxn_high = get_fa_scorefxn() #  create_score_function('standard', 'score12')\n","\n","    # 11. setup a RepeatMover on a TrialMover of a SequenceMover\n","    # -setup a TrialMover\n","    #    a. create a SequenceMover of the fragment insertions\n","    #### add any other moves you desire\n","    folding_mover = protocols.moves.SequenceMover()\n","    folding_mover.add_mover(insert_long_frag)\n","    folding_mover.add_mover(insert_short_frag)\n","    #    b. create a MonteCarlo object to define success/failure\n","    # must reset the MonteCarlo object for each trajectory!\n","    mc = MonteCarlo(test_pose, scorefxn_low, kT)\n","    # c. create the TrialMover\n","    trial = TrialMover(folding_mover, mc)\n","\n","    #### for each trajectory, try cycles number of applications\n","\n","    # -create the RepeatMover\n","    folding = protocols.moves.RepeatMover(trial, cycles)\n","\n","    # 12. create a (Py)JobDistributor\n","    jd = PyJobDistributor(job_output, jobs, scorefxn_high)\n","\n","    # 13. store the score evaluations for output\n","    # printing the scores as they are produced would be difficult to read,\n","    #    Rosetta produces a lot of verbose output when running\n","    scores = [0]*(jobs + 1)\n","    scores[0] = scorefxn_low(pose)\n","\n","    # 14. perform folding by\n","    counter = 0    # for exporting to PyMOL\n","    while not jd.job_complete:\n","        # a. set necessary variables for the new trajectory\n","        # -reload the starting pose\n","        test_pose.assign(pose)\n","        # -change the pose's PDBInfo.name, for the PyMOL_Observer\n","        counter += 1\n","        test_pose.pdb_info().name(job_output + '_' + str(counter))\n","        # -reset the MonteCarlo object (sets lowest_score to that of test_pose)\n","        mc.reset(test_pose)\n","\n","        #### if you create a custom protocol, you may have additional\n","        ####    variables to reset, such as kT\n","\n","        #### if you create a custom protocol, this section will most likely\n","        ####    change, many protocols exist as single Movers or can be\n","        ####    chained together in a sequence (see above) so you need\n","        ####    only apply the final Mover\n","        # b. apply the refinement protocol\n","        folding.apply(test_pose)\n","\n","        ####\n","        # c. export the lowest scoring decoy structure for this trajectory\n","        # -recover the lowest scoring decoy structure\n","        mc.recover_low(test_pose)\n","        # -store the final score for this trajectory\n","        scores[counter] = scorefxn_low(test_pose)\n","        # -convert the decoy to fullatom\n","        # the sidechain conformations will all be default,\n","        #    normally, the decoys would NOT be converted to fullatom before\n","        #    writing them to PDB (since a large number of trajectories would\n","        #    be considered and their fullatom score are unnecessary)\n","        # here the fullatom mode is reproduced to make the output easier to\n","        #    understand and manipulate, PyRosetta can load in PDB files of\n","        #    centroid structures, however you must convert to fullatom for\n","        #    nearly any other application\n","        to_fullatom.apply(test_pose)\n","        # -guess what cysteines are involved in disulfide bridges\n","        guess_disulfides(test_pose)\n","        # -output the fullatom decoy structure into a PDB file\n","        jd.output_decoy(test_pose)\n","        # -export the final structure to PyMOL\n","        test_pose.pdb_info().name(job_output + '_' + str(counter) + '_fa')\n","\n","        #### if you want to see the decoy scores, uncomment the line below\n","        #scorefxn_high( test_pose )\n","\n","    # 15. output the score evaluations\n","    print( '===== Centroid Scores =====' )\n","    print( 'Original Score\\t:\\t', scores[0] )\n","    for i in range(1, len(scores)):    # print out the job scores\n","        # the \"[:14].ljust(14)\" is to force the text alignment\n","        print( (job_output + '_' + str( i ))[:14].ljust(14) +\\\n","            '\\t:\\t', scores[i] )\n","\n","    return scores    # for other protocols\n","\n","# locates all cysteines, make bonds between them, output the bonds that lower\n","#    the score\n","def guess_disulfides(pose, cutoff = 6.0):\n","    \"\"\"\n","    A quick method for probing a protein for cysteine residues close to each\n","        other (within  <cutoff>  )\n","    \"\"\"\n","    # find all cysteine residues and consider possible disulfides\n","    cys = [i for i in range(1, pose.total_residue() + 1)\n","        if pose.residue(i).name1() == 'C']\n","    partners = [0]*sum( range(len(cys)) )    # all disulfides possible\n","    i = 0\n","    # create all combinations\n","    for first in range(len(cys[:-1])):\n","        for second in cys[first + 1:]:\n","            partners[i] = (cys[first], second)\n","            i += 1\n","    # try each disulfide, if it lowers the score, print it to screen\n","    print( '='*80 )\n","    print( 'Potential Disulfides:' )\n","    for pair in partners:\n","        # for a fullatom cysteine in PyRosetta, the 6th atom is sulfur\n","        separation = (pose.residue(pair[0]).xyz(6) -\n","            pose.residue(pair[1]).xyz(6)).norm\n","        if separation < cutoff:\n","            print( 'between (pose numbered) residue' , pair[0] , 'and' ,\\\n","                pair[1] , '|' , separation , 'Angstrom separation' )\n","    print( '='*80 )\n","    # to manipulate disulfide bonds, use:\n","    # formation:    form_disulfide(pose.conformation(), 6, 16)\n","    # cleavage:     change_cys_state(6, 'CYS', pose.conformation() )\n","    #               change_cys_state(16, 'CYS', pose.conformation() )\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"7tmLHsk5xAeT"},"source":["## Here we define input variables"]},{"cell_type":"markdown","metadata":{"id":"fpfuN_HVygkL"},"source":["\n","-   The long_frag_length MUST match long_frag_filename (9 in this example)         \n","\n","-   The short_frag_length MUST match short_frag_filename (3 here)\n","-   100 trajectories is low, sampling protein conformations requires many trials, typically hundreds (800-1000) trajectories are attempted         \n","-   The \"temperature\" parameter, kT, is set to 1.0, a neutral value, larger kT increase the diversity of sampling (easily escape local minima) but requires compensation with more trajectories smaller kT decreases the chance of sampling \"useless\" space but cannot easily escape local minima (the default kT = 3.0)\n","- There are no common values for long_inserts, short_inserts, or cycles more inserts indicates greater sampling with fragments while greater cycles produces more sampling in general (though in this case only fragment insertion is performed) however, only one selection (Metropolis Criteria) is applied per cycle, thus a proposed structure with too many fragments inserted is likely to be rejected but additional trials increases the computation time noticeably here, as is typical is Rosetta, a fairly conservative change is proposed (1 9-mer and 3 3-mers) per cycle with many cycles"]},{"cell_type":"code","metadata":{"id":"qPBVkUh0w9Yr"},"source":["# the user may input a PDB file, fasta file, or sequence directly\n","# PDB file option\n","pose = Pose()\n","# Fasta file option\n","fasta_filename = './t000_.fasta'\n","if fasta_filename:    # defaults to off, empty string\n","    f = open(fasta_filename, 'r')    # open the file\n","    sequence = f.readlines()    # read the text\n","    f.close()    # close it\n","    # removing the trailing \"\\n\" and any header lines\n","    sequence = [line.strip() for line in sequence if not '>' in line]\n","    sequence = ''.join( sequence )    # combine into a single sequence\n","elif options.sequence:\n","    sequence = options.sequence\n","#else:\n"," #   pose_from_file(pose, pdb_filename)\n"," #   sequence = pose.sequence()\n","#Checks for the sequence in a fasta, then direct, and finally from a PDB file.  If no PDB file is given, it will load the default.\n","\n","# fragment files options\n","long_frag_filename = './aat000_09_05.200_v1_3'\n","long_frag_length = 9\n","short_frag_filename = './aat000_03_05.200_v1_3'\n","short_frag_length = 3\n","# folding protocol options, see above\n","kT = 3.0\n","long_inserts = 1\n","short_inserts = 3\n","cycles = 300\n","# PyJobDistributor options\n","jobs = 10\n","job_output = './outs/fold_output'                  #Prefix"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"E2yCqcQdxEvc"},"source":["## Here we finally call the folding function"]},{"cell_type":"code","metadata":{"id":"111jvH6XxF0x"},"source":["sample_folding(sequence,\n","    long_frag_filename, long_frag_length,\n","    short_frag_filename, short_frag_length,\n","    kT, long_inserts, short_inserts, cycles,\n","    jobs, job_output)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"QOAPtmkSePwn"},"source":["# Part III – Analyze the results"]},{"cell_type":"markdown","metadata":{"id":"LXBRnThEeyhq"},"source":["As a result of running Rosetta, two things will happen:\n","1. All the 10 predicted structures for villin headpiece will be outputted as PDB files\n","2. An energy score and statistics file of the output, called score.fsc, will also be continually generated"]},{"cell_type":"code","metadata":{"id":"Npex-swMLPCl"},"source":["!sudo apt install jq"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"43NBfnMGNbyl"},"source":["#As we obtain a Json scorefile we use the jq bash json parser to filter the decoy name and total_score\n","#in order to get the lowest energy model\n","!jq '.decoy , .total_score' /content/outs/fold_output.fasc"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"0Y4qwnirfkHN"},"source":["Now is your turn to the 2PPZ(native.pdb) pdb and align it to the structure with the predicted structure with the lowest energy using py3Dmol and BioPython!!"]},{"cell_type":"code","metadata":{"id":"85ZdGjTZIaCf"},"source":["!pip install biopython\n","#We import Bio.PDB module again\n","from Bio.PDB import *\n","parser = PDBParser()"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"bUwfwJaxIblz"},"source":["#The following code was created by Anders Steen Christensen\n","#from the University of Basel and is available at\n","#https://gist.github.com/andersx/6354971\n","\n","import Bio.PDB\n","\n","# Select what residues numbers you wish to align\n","# and put them in a list\n","start_id = 1\n","end_id   = 35\n","atoms_to_be_aligned = range(start_id, end_id + 1)\n","\n","# Start the parser\n","pdb_parser = Bio.PDB.PDBParser(QUIET = True)\n","\n","# Get the structures\n","ref_structure = pdb_parser.get_structure(\"reference\", \"/content/native.pdb\")\n","sample_structure = pdb_parser.get_structure(\"sample\", \"/content/outs/fold_output_4.pdb\")\n","\n","# Use the first model in the pdb-files for alignment\n","# Change the number 0 if you want to align to another structure\n","ref_model    = ref_structure[0]\n","sample_model = sample_structure[0]\n","\n","# Make a list of the atoms (in the structures) you wish to align.\n","# In this case we use CA atoms whose index is in the specified range\n","ref_atoms = []\n","sample_atoms = []\n","\n","# Iterate of all chains in the model in order to find all residues\n","for ref_chain in ref_model:\n","  # Iterate of all residues in each model in order to find proper atoms\n","  for ref_res in ref_chain:\n","    # Check if residue number ( .get_id() ) is in the list\n","    if ref_res.get_id()[1] in atoms_to_be_aligned:\n","      # Append CA atom to list\n","      ref_atoms.append(ref_res['CA'])\n","\n","# Do the same for the sample structure\n","for sample_chain in sample_model:\n","  for sample_res in sample_chain:\n","    if sample_res.get_id()[1] in atoms_to_be_aligned:\n","      sample_atoms.append(sample_res['CA'])\n","\n","# Now we initiate the superimposer:\n","super_imposer = Bio.PDB.Superimposer()\n","super_imposer.set_atoms(ref_atoms, sample_atoms)\n","super_imposer.apply(sample_model.get_atoms())\n","\n","# Print RMSD:\n","print('The calculated RMSD is:')\n","print (super_imposer.rms)\n","\n","# Save the aligned version of one of the chains of 6ANE\n","io = Bio.PDB.PDBIO()\n","io.set_structure(sample_structure)\n","io.save(\"Your_best_model_aligned.pdb\")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"ny5zIv8wP6hm"},"source":["#Here, write down the code to install py3Dmol\n","!pip install py3Dmol"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"JcFXtv33f64Z"},"source":["#And we already learned how to import these modules, right?\n","import py3Dmol\n","view=py3Dmol.view()\n","view.addModel(open('/content/Your_best_model_aligned.pdb', 'r').read(),'pdb')\n","view.addModel(open('/content/native.pdb', 'r').read(),'pdb')\n","view.zoomTo()\n","view.setBackgroundColor('white')\n","view.setStyle({'chain':'A'},{'cartoon': {'color':'yellow'}})\n","view.setStyle({'chain':'B'},{'cartoon': {'color':'green'}})\n","view.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"rSxJGITc1tBU"},"source":["# Appendix A. Going deeper"]},{"cell_type":"markdown","metadata":{"id":"f4b0gE4GmAx9"},"source":["## A Real Example\n","\"\"\"\n","All of the default variables and parameters used above are specific to\n","the example with \"test_in.pdb\", which is supposed to be simple,\n","straightforward, and speedy. Here is a more practical example:\n","\n","Pheromone ER-23 is a small protein (51 amino acids) with 5 disulfide bridges.\n","The formation of these bonds may significantly effect folding. Suppose you\n","are interested in the significance of these disulfide bonds as folding\n","constraints and want to predict the protein without them using PyRosetta.\n","\n","1. Obtain the protein sequence of RCSB PDB 1HA8 (instructions above)\n","        (for the example here, download the PDB file and ensure it is clean)\n","2. Make two fragment files using the \"Generate Fragment Files\"\n","        (instructions above)\n","        -of 9-mers (the long fragments)\n","        -of 3-mers (the short fragments)\n","3. Make a directory containing:\n","        -the PDB file for 1HA8 (cleaned of HETATMs and waters)\n","            lets name it \"1HA8.clean.pdb\" here\n","        -the 9-mer fragment file for 1HA8\n","            lets name it \"1HA8.frag9\" here\n","        -the 3-mer fragment file for 1HA8\n","            lets name it \"1HA8.frag3\" here\n","        -this sample script (technically not required, but otherwise the\n","            commands in 4. would change since folding.py would not be here)\n","4. Run the script from the commandline with appropriate arguments:\n","\n",">python folding.py --pdb_filename 1HA8.clean.pdb --long_frag_filename 1HA8.frag9 --long_frag_length 9 --short_frag_filename 1HA8.frag3 --short_frag_length 3 --jobs 100 --job_output 1HA8_folding_output --kT 1.0 --long_inserts 1 --short_inserts 3 --cycles 200 --disulfides 1\n","\n","        -The long_frag_length MUST match long_frag_filename (9 in this example)\n","        -The short_frag_length MUST match short_frag_filename (3 here)\n","        -100 trajectories is low, sampling protein conformations requires many\n","            trials, typically hundreds (800-1000) trajectories are attempted\n","        -The \"temperature\" parameter, kT, is set to 1.0, a neutral value,\n","            larger kT increase the diversity of sampling (easily escape local\n","            minima) but requires compensation with more trajectories\n","            smaller kT decreases the chance of sampling \"useless\" space but\n","            cannot easily escape local minima (the default kT = 3.0\n","        -There are no common values for long_inserts, short_inserts, or cycles\n","            more inserts indicates greater sampling with fragments while\n","            greater cycles produces more sampling in general (though in this\n","            case only fragment insertion is performed)\n","            however, only one selection (Metropolis Criteria) is applied per\n","            cycle, thus a proposed structure with too many fragments inserted\n","            is likely to be rejected but additional trials increases the\n","            computation time noticeably\n","            here, as is typical is Rosetta, a fairly conservative change is\n","            proposed (1 9-mer and 3 3-mers) per cycle with many cycles\n","\n","5. Wait for output, this will take a while (performing 100 trajectories\n","        of loop modeling involving 1*3*200 total moves per trajectory)\n","6. Analyze the results (see INTERPRETING RESULTS above)\n","\n","Note: this is NOT intended to be used for realistic centroid folding,\n","it merely provides a \"skeleton\" for the code in PyRosetta. It may be useful\n","for preliminary investigation but the best protocols are somewhat\n","protein-specific, there is no current universal folding method. As mentioned\n","above, this script only touches the low-resolution aspect of Rosetta folding\n","and must be accompanied by a high-resolution step to yield complete results.\n","Generally, a protocol similar to the one presented here with more drastic\n","sampling and a larger number of trials should be sufficient to find realistic\n","folds (after refinement).\n","\n","\"\"\"\n","\n"]},{"cell_type":"markdown","metadata":{"id":"IfjVIjZ7aFIJ"},"source":["\n","## Changing Folding Sampling\n","\"\"\"\n","Rosetta is a structure prediction tool and all predictions rely on the\n","techniques (moves) used to generate proposed structures. There is NO GENERAL\n","SOLUTION for predicting a protein fold from scratch and fragment insertion\n","is an effective way of sampling the relevant space. Since fragments are\n","sequence and secondary structure (prediction) dependent, the resultant folds\n","are biochemically feasible and in accordance with our prior observations about\n","protein stucture (the PDB).\n","\n","For speed, Rosetta folding (and the folding algorithm presented here) is\n","performed in low-resolution (centroid) mode. This method of folding is\n","abstractly looking for structures which obey certain size constraints (i.e. the\n","rg score term) but do not unrealistically clash (i.e. the vdw score term) that\n","are achieved easily (i.e. fragment insertion, a fast algorithm). Sufficient\n","sampling and downstream refinement will convert these general folds (the output\n","of this script) into reasonable estimates of the native protein structure.\n","\n","Many other sampling techniques and methods are available. PyRosetta was designed\n","to make novel investigation of these ideas easy. Recommend literature include\n","topics in protein structure prediction, Monte Carlo Markov Chain processes,\n","Rosetta protocols, and simulated annealing algorithms. When developing custom\n","protocols, you MUST understand the Movers available in PyRosetta (see the other\n","sample scripts, specifically refinement.py). Unfortunately, you cannot build\n","novel Mover classes in PyRosetta. Fortunately, it is easy to implement similar\\\n","code on your own once you understand how moves and selection are performed.\n","\n","Please try alternate sampling methods to better understand how these\n","algorithms perform and to find what moves best suite your problem.\n","\n","\"\"\"\n","\n"]},{"cell_type":"markdown","metadata":{"id":"AjaMrOtDaH0h"},"source":["\n","## Changing Folding Scoring\n","\"\"\"\n","The Rosetta \"force-field\" is a biasing function (see MCMC algorithms) used to\n","predict protein structures based on what is observed about other protein\n","structures. These scoring methods supply the constraints necessary to bias\n","structure prediction but incorporate a LARGE number of fine tuned parameters.\n","There is no easy way to optimize all relevant constants and performance has been\n","increased by balancing speed and accuracy.\n","\n","Many other relevant metrics exist for protein structures and discriminating\n","decoys (incorrect predictions) from actual proteins is an ongoing task. Please\n","experiment with different ScoreTypes in PyRosetta, their weights, and the\n","combination of terms. Unfortunately, you cannot build novel ScoreTypes (scoring\n","methods) in PyRosetta. Fortunately, it is easy to implement similar code on your\n","own once you understand how scoring biases MCMC simulations.\n","\n","Please try alternate scoring functions or unique selection methods to better\n","understand which scoring terms contribute to performance and to find what\n","scoring best suites your problem.\n","\n","\"\"\""]}]}
