{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"gDvK9xHtcjTr"},"source":["#Lab.05 / IBM3202 ‚Äì¬†Modelamiento de Prote√≠nas de Membrana usando Rosetta"]},{"cell_type":"markdown","metadata":{"id":"CTFZC7Mafuzu"},"source":["##Aspectos Te√≥ricos"]},{"cell_type":"markdown","metadata":{"id":"VWTkCQifjWXY"},"source":["Seg√∫n Alford et al. 2020 [Biophys J 2020 118(8), 2042-2055], las prote√≠nas de membrana colaboran con el entorno lip√≠dico circundante para llevar a cabo procesos vitales esenciales. Constituyen el 30% de todas las prote√≠nas y son objetivos de m√°s del 60% de los productos farmac√©uticos. Sin embargo, las dificultades experimentales han limitado nuestra comprensi√≥n de sus mecanismos moleculares de funcionamiento.\n"]},{"cell_type":"markdown","metadata":{"id":"0nPD5ByvctQy"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/mpmodelling_01.jpg'/>\n","<figcaption>FIGURA 1. El desaf√≠o de resolver la estructura de prote√≠nas de membrana. Como se muestra en este gr√°fico, el n√∫mero de prote√≠nas de membrana resueltas antes del amplio uso de la criomicroscop√≠a electr√≥nica (cryoEM) era (y sigue siendo) significativamente inferior al n√∫mero de prote√≠nas solubles.\n","\n","<br>Bijelic A & Rompel A (2015) <i>Coord Chem Rev 299, 22-38.\n","</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"iHpJP9wCdgqp"},"source":["Por lo tanto, la modelizaci√≥n de prote√≠nas de membrana es ideal para determinar la estructura de estas prote√≠nas y para desarrollar nuevos tratamientos terap√©uticos. En este tutorial, aprenderemos a modelar prote√≠nas de membrana en PyRosetta y proporcionaremos ejemplos de los tipos de problemas que se pueden resolver con este marco de trabajo.\n"]},{"cell_type":"markdown","metadata":{"id":"wr-vcvaVTKTf"},"source":["##Generalidades"]},{"cell_type":"markdown","metadata":{"id":"62vLxb1sSVVf"},"source":["En este tutorial, modelaremos diferentes hom√≥logos con identidades de secuencia variables de la prote√≠na de membrana Glicoforina A, tambi√©n conocida como GYPA. Esta es una prote√≠na formadora de poros que, en el caso del ort√≥logo humano, ha sido ampliamente estudiada mediante diferentes m√©todos de biolog√≠a estructural, como la resonancia magn√©tica nuclear (NMR) y la cristalograf√≠a de rayos X.\n","\n","Esta visi√≥n general y tutorial incluyen informaci√≥n de este [tutorial](http://www.meilerlab.org/jobs/downloadfile/name/May2018_rosetta_cm_presentation.pdf) por Nica Marlow y Georg Kuenze.\n"]},{"cell_type":"markdown","metadata":{"id":"jmsmt7SGbTmQ"},"source":["**El Algoritmo**\n","\n","En t√©rminos generales, el algoritmo consiste en una larga trayectoria de Monte Carlo que comienza desde una plantilla elegida al azar. La trayectoria de Monte Carlo utiliza los siguientes movimientos:\n","\n","* Inserci√≥n de fragmentos en regiones no alineadas.\n","\n","* Reemplazo de un segmento elegido al azar con el de una estructura de plantilla diferente.\n","\n","* Minimizaci√≥n en el espacio cartesiano utilizando una versi√≥n suave (diferenciable) de la funci√≥n de energ√≠a del centroide de Rosetta. Finalmente, esto se sigue de una optimizaci√≥n a nivel de √°tomos.\n","\n","Consulta m√°s en la [documentaci√≥n oficial](https://www.rosettacommons.org/docs/latest/application_documentation/structure_prediction/RosettaCM).\n"]},{"cell_type":"markdown","metadata":{"id":"_rm1_EjsTMWF"},"source":["**Diferencias entre Modelado con un Solo Molde y Modelado con M√∫ltiples Moldes**\n","\n","*Modelado con un Solo Molde:*\n","- Se utiliza un solo molde como entrada.\n","- Emplea fragmentos derivados de la secuencia y el molde.\n","- Se utiliza cuando los moldes disponibles tienen una identidad muy alta (>60%).\n","\n","*Modelado con M√∫ltiples Moldes:*\n","- Se utilizan m√∫ltiples moldes como entrada.\n","- Combina secciones de m√∫ltiples modelos enhebrados y fragmentos derivados de la secuencia.\n","- Se utiliza cuando los moldes disponibles tienen una baja identidad (30-50%).\n"]},{"cell_type":"markdown","metadata":{"id":"eiINEFYXfajF"},"source":["**El protocolo**\n","\n","El modelado comparativo de Rosetta (**RosettaCM**) realiza una serie compleja de pasos que se ilustran en el esquema a continuaci√≥n, pero que se pueden resumir en tres pasos principales:\n","\n","1. Alinear la secuencia objetivo con las secuencias de los moldes.\n","\n","2. Realizar un enhebrado parcial de la secuencia objetivo en las estructuras de los moldes.\n","\n","3. Combinar fragmentos de diferentes moldes utilizando la hibridaci√≥n de RosettaCM.\n"]},{"cell_type":"markdown","metadata":{"id":"NK5aEeL_e2w_"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/mpmodelling_02.jpg'/>\n","<figcaption>FIGURA 2. Un esquema simple del protocolo RosettaCM\n","<br>Song Y et al (2013) <i>Structure 21(10), 1735-1742</i></figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"eTZzTm6SUOIH"},"source":["**Archivos de Entrada para RosettaCM**\n","\n","M√≠nimo indispensable:\n","\n","* Estructuras enhebradas parcialmente (‚úÖ)\n","\n","* Definici√≥n del movimiento y opciones (‚úÖ)\n","\n","Espec√≠ficos para prote√≠nas de membrana (no necesarios si se modelan prote√≠nas solubles):\n","\n","* Regiones que atraviesan la membrana (archivo \"span\") (üôã)\n","\n","* Parches de ponderaci√≥n para la membrana (‚úÖ)\n","\n","Archivos opcionales basados en la informaci√≥n disponible:\n","\n","* Informaci√≥n de restricciones (por ejemplo, conectividad de pares de √°tomos)\n","\n","* Conectividad de disulfuros\n"]},{"cell_type":"markdown","metadata":{"id":"SKd7eVaiRunj"},"source":["**Para una exploraci√≥n profunda del modelado de prote√≠nas de membrana con Rosetta, sugerimos revisar los siguientes art√≠culos:**\n","1. Alford RF, Leman JK et al., ‚ÄúAn Integrated Framework Advancing Membrane Protein Modeling and Design,‚Äù PLOS Computational Biology. 2015;11(9).\n","2. Alford RF, Fleming PJ, Fleming KG, Gray JJ (2020) ‚ÄúProtein structure prediction and design in a biologically-realistic implicit membrane‚Äù BioRxiv\n","3. Koehler Leman J, Muller BK, Gray JJ (2017) ‚ÄúExpanding the toolkit for membrane protein modeling in Rosetta‚Äù Bioinformatics 33(5):754-756.\n","4. Koehler Leman J & Gray JJ (2015) ‚ÄúComputational modeling of membrane proteins‚Äù Proteins Structure, Function, and Bioinformatics 83(1):1-24."]},{"cell_type":"markdown","metadata":{"id":"U2vWpWg8klhM"},"source":["#Parte I ‚Äì Explorando las secuencias de los ort√≥logos GYPA"]},{"cell_type":"markdown","metadata":{"id":"D-u6sgce8Zxm"},"source":["1. Las secuencias con las que trabajaremos hoy corresponden a tres ort√≥logos de la prote√≠na de membrana Glicoforina A. En nuestro caso particular, modelaremos los siguientes ort√≥logos utilizando PyRosetta:\n","\n","  *   [Marmota flaviventris](https://www.ncbi.nlm.nih.gov/protein/XP_027782564.1?report=fasta&log$=seqview)\n","\n","  *   [Gorilla gorilla](https://www.ncbi.nlm.nih.gov/protein/XP_007666620.2?report=fasta&log$=seqview)\n","  \n","  *   [Ornithorhynchus anatinus](https://www.uniprot.org/uniprot/O18798)"]},{"cell_type":"markdown","metadata":{"id":"xHLARFm0-8ae"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/mpmodelling_03.png'/>\n","</figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"ESDFFpOd-pYU"},"source":["2. Por cuestiones de tiempo, proporcionaremos las secuencias ya alineadas y recortadas en un archivo en formato Grishin (otro formato de alineaci√≥n), las estructuras enhebradas parcialmente necesarias para ejecutar el protocolo de modelado y los archivos de fragmentos:\n","\n","   Encontrar√°s estos archivos dentro de su carpeta correspondiente en input_files. Si deseas generar estos archivos de entrada por ti mismo, consulta **Ap√©ndice A**.\n"]},{"cell_type":"code","metadata":{"id":"ImD_xOcegVuQ"},"source":["#Descargamos los input desde Github\n","!wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/input_files/input_files_mp.zip\n","#Comando para eliminar en caso de que hayan archivos anteriores\n","!rm -rf input_files\n","#Descomprimir los archivos\n","!unzip input_files_mp.zip"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"agqXm_0dmf8k"},"source":["#Parte II ‚Äì Determinando las regiones que atraviesan membrana (span file) usando OCTOPUS server"]},{"cell_type":"markdown","metadata":{"id":"a0Qhw-MSmtjk"},"source":["Como mencionamos brevemente al principio de este tutorial (Visi√≥n General), necesitamos proporcionarle a Rosetta una manera de saber a qu√© residuos se debe aplicar un \"potencial de membrana\" espec√≠fico e impl√≠cito. Estas instrucciones se generan en base al servidor de predicci√≥n de regiones transmembrana (TM) [**OCTOPUS**](https://octopus.cbr.su.se).\n"]},{"cell_type":"markdown","metadata":{"id":"ZIycpMeaCwI4"},"source":["1. En [**OCTOPUS**](https://octopus.cbr.su.se), debes ingresar la secuencia de la prote√≠na objetivo para la predicci√≥n de regiones transmembrana (TM). Puedes encontrar estas secuencias de prote√≠nas dentro de la carpeta FASTA en los archivos de entrada descargados.\n"]},{"cell_type":"markdown","metadata":{"id":"jNyDLK3CCShq"},"source":["2. Luego, los archivos de topolog√≠a deben procesarse para convertirlos en archivos \"span\", como se muestra en la celda de c√≥digo a continuaci√≥n:"]},{"cell_type":"code","metadata":{"id":"u_pxQELHbPX7"},"source":["%%bash\n","cd /content/input_files/octopus\n","perl octopus2span.pl GYPA.topo.txt > GYPA.span\n","perl octopus2span.pl marmota.topo.txt > marmota.span\n","perl octopus2span.pl ornithorhynchus.topo.txt > ornithorhynchus.span\n","perl octopus2span.pl gorilla.topo.txt > gorilla.span"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zjUSFLdJCY7f"},"source":["3. Una vez que hayas obtenido todos los archivos \"span\" de **OCTOPUS**, recuerda colocarlos dentro de la siguiente carpeta:\n","\n","\n","```\n","/content/input_files/octopus #Puedes abrir la carpeta aprentando esta direcci√≥n\n","```\n","\n"]},{"cell_type":"markdown","metadata":{"id":"o3hcWpFQXoUx"},"source":["# Parte III ‚Äì Instalando PyRosetta y ejecutando los protocolos de modelamiento de prote√≠nas de membrana"]},{"cell_type":"markdown","metadata":{"id":"i4mcFtHR6u_-"},"source":["1. Comenzamos configurando PyRosetta en Google Colab, bas√°ndonos en la versi√≥n de PyRosetta que has compilado e instalado previamente y que est√° disponible en tu Google Drive.\n","\n","**‚ö†Ô∏èADVERTENCIA:** ¬°Recuerda que DEBES haber compilado e instalado PyRosetta antes de seguir este tutorial!\n"]},{"cell_type":"code","metadata":{"id":"SePtDf1JST8a"},"source":["!pip install pyrosettacolabsetup\n","import pyrosettacolabsetup\n","pyrosettacolabsetup.install_pyrosetta_on_google_drive()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"MglIIrmAhWZ0"},"source":["2. En la siguiente celda de c√≥digo, importamos los m√≥dulos de PyRosetta que vamos a utilizar:\n"]},{"cell_type":"code","metadata":{"id":"e02ruAYR076e"},"source":["#Importamos los m√≥dulos de pyRosetta que utilizaremos\n","%%time\n","import pandas as pd\n","from pyrosetta import *\n","from pyrosetta.rosetta.protocols.rosetta_scripts import *\n","from pyrosetta import (\n","    init, pose_from_sequence, pose_from_file, Pose, MoveMap, create_score_function, get_fa_scorefxn,\n","    MonteCarlo, TrialMover, SwitchResidueTypeSetMover, PyJobDistributor,\n",")\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"4vFqicSeD-NQ"},"source":["3. A continuaci√≥n, llevamos a cabo el protocolo de Modelado de Prote√≠nas de Membrana. Esto se describe brevemente para el GYPA de _Gorilla_, mientras que el modelado del GYPA para los otros ort√≥logos es simplemente una modificaci√≥n de este protocolo.\n"]},{"cell_type":"markdown","metadata":{"id":"wlmDdcxmXZ9T"},"source":["#### Gorilla ü¶ç"]},{"cell_type":"markdown","metadata":{"id":"cQc4_jK8Oug2"},"source":["1. En la siguiente celda de c√≥digo, proporcionamos a PyRosetta todas las **Flags** necesarias. Estas flags son equivalentes a instrucciones de l√≠nea de comandos para establecer la ruta de los archivos y configurar el valor de diversas opciones (numerosas). Las pasamos a PyRosetta, inicializ√°ndolo de esta manera.\n","\n","\n","\n","```\n","pyrosetta.init(flags)\n","```\n","\n"]},{"cell_type":"code","metadata":{"id":"vG9dJLT4bZkk"},"source":["#Las 'flags' son como las opciones del c√≥digo en l√≠nea de comando\n","flags = f\"\"\"\n","-no_fconfig\n","-in:file:fasta ./input_files/fasta/gorillaGYPA.fasta\n","-out:prefix GYPA_gorilla_\n","-out:path:all ./output_files\n","-in:file:spanfile ./input_files/octopus/gorilla.span\n","-membrane:no_interpolate_Mpair\n","-membrane:Menv_penalties\n","-rg_reweight .1\n","-relax:fast\n","-relax:minimize_bond_angles\n","-relax:minimize_bond_lengths\n","-relax:jump_move true\n","-default_max_cycles 2\n","-relax:min_type lbfgs_armijo_nonmonotone\n","-score:weights input_files/wts/stage3_rlx_membrane.wts\n","-use_bicubic_interpolation\n","-hybridize:stage1_probability 1.0\n","-sog_upper_bound 15\n","-chemical:exclude_patches LowerDNA  UpperDNA Cterm_amidation SpecialRotamer VirtualBB ShoveBB VirtualDNAPhosphate VirtualNTerm CTermConnect sc_orbitals pro_hydroxylated_case1 pro_hydroxylated_case2 ser_phosphorylated thr_phosphorylated  tyr_phosphorylated tyr_sulfated lys_dimethylated lys_monomethylated  lys_trimethylated lys_acetylated glu_carboxylated cys_acetylated tyr_diiodinated N_acetylated C_methylamidated MethylatedProteinCterm\n","-analytic_etable_evaluation true\n","-linmem_ig 10\n","-out:mute core.io.pose_from_sfr.PoseFromSFRBuilder core.scoring.MembranePotential core.chemical.GlobalResidueTypeSet basic.io.database core.scoring.MembraneTopology core.scoring.CartesianBondedEnergy core.conformation.Conformation protocols.jd2.parser.ScoreFunctionLoader core.optimization.AtomTreeMinimizer core.pack\n","\"\"\"\n","pyrosetta.init(flags)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"3EwFD1YdO-xV"},"source":["2. Uno de los conceptos cruciales en PyRosetta son las **poses**. Contienen la informaci√≥n, ya sea secuencia o estructura, a la cual se aplicar√° el protocolo.\n","\n","   Teniendo esto en cuenta, aqu√≠ creamos un **objeto pose** a partir de una secuencia, que en este caso es la secuencia del GYPA de **_Gorilla_**. Tambi√©n especificamos un **res_type** que corresponde a **fullatom_standard**.\n"]},{"cell_type":"code","metadata":{"id":"MQRxSQh4bZku"},"source":["#Aqu√≠ creamos un objeto que posea la secuencia\n","#tambi√©n especificamos un res_type que corresponde a fullatom_standard\n","%%time\n","pose = pose_from_sequence('DNGEWIQLVPRFSGPEITLIIFGVMAGIIGTILLISYSIRRL', 'fa_standard')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"40X9c4WjPXeq"},"source":["3. Aunque hay m√∫ltiples formas de crear un protocolo en Rosetta, en nuestro caso vamos a crear un protocolo a partir de un **objeto XML**.\n","\n","**NOTA:** ¬°En la etiqueta **MOVERS** especificamos los **archivos de fragmentos** y el **archivo threaded_pdb**!\n"]},{"cell_type":"code","metadata":{"id":"pLfWzrs0bZky"},"source":["#Aqu√≠ creamos un protocolo de Rosetta usando c√≥digo XML\n","#Nota que en <MOVERS> especificamos los fragmentos y el PDB enhebrado!!\n","\n","%%time\n","xml = pyrosetta.rosetta.protocols.rosetta_scripts.XmlObjects.create_from_string(\"\"\"\n","<ROSETTASCRIPTS>\n","    <TASKOPERATIONS>\n","    </TASKOPERATIONS>\n","    <RESIDUE_SELECTORS>\n","    </RESIDUE_SELECTORS>\n","    <SCOREFXNS>\n","      <ScoreFunction name=\"stage1\" weights=\"input_files/wts/stage1_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"1\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"stage2\" weights=\"input_files/wts/stage2_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"fullatom\" weights=\"input_files/wts/stage3_rlx_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"membrane\" weights=\"membrane_highres_Menv_smooth\" symmetric=\"0\">\n","        <Reweight scoretype=\"cart_bonded\" weight=\"0.5\"/>\n","        <Reweight scoretype=\"pro_close\" weight=\"0\"/>\n","\t <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","\t</ScoreFunction>\n","    </SCOREFXNS>\n","    <FILTERS>\n","    </FILTERS>\n","    <MOVERS>\n","        \t <Hybridize name=\"hybridize\" stage1_scorefxn=\"stage1\" stage2_scorefxn=\"stage2\" fa_scorefxn=\"fullatom\" batch=\"1\" stage1_increase_cycles=\"0.5\" stage2_increase_cycles=\"0.5\" linmin_only=\"1\" add_hetatm=\"0\" hetatm_to_protein_cst_weight=\"0\"  >\n","       \t\t\t<Fragments three_mers=\"/content/input_files/frags/gorillaGYPA.9mers\" nine_mers=\"/content/input_files/frags/gorillaGYPA.9mers\" />\n","\t          \t<Template pdb=\"/content/input_files/threaded_pdbs/5eh6_A.gorilla.pdb\" cst_file=\"AUTO\" weight=\"1.000\" />\n","\t       </Hybridize>\n","    </MOVERS>\n","    <APPLY_TO_POSE>\n","    </APPLY_TO_POSE>\n","    <PROTOCOLS>\n","      Add mover=\"add_memb\"\n","\t    <Add mover=\"hybridize\"/>\n","    </PROTOCOLS>\n","<OUTPUT scorefxn=\"membrane\"/>\n","</ROSETTASCRIPTS>\n","    \"\"\").get_mover(\"ParsedProtocol\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Dn9Gu8TkjmVU"},"source":["4. Una vez que nuestro protocolo est√° configurado, en las siguientes celdas de c√≥digo:\n","   * Creamos una funci√≥n de puntuaci√≥n para membranas.\n","   * Luego, llamamos al PyJobDistributor y pasamos tres argumentos:\n","     * Un prefijo de nombre para guardar nuestros modelos como archivos PDB.\n","     * El n√∫mero de modelos a generar.\n","     * La funci√≥n de puntuaci√≥n a emplear.\n","   * Pasamos nuestra pose al distribuidor de trabajos que acabamos de crear.\n","   * Despu√©s, creamos un DataFrame de Pandas para almacenar los resultados de las puntuaciones.\n","   * Finalmente, iteramos mientras los trabajos no est√©n completados.\n"]},{"cell_type":"code","metadata":{"id":"flBA8JKLbZk6"},"source":["%%time\n","#Llamamos a la funci√≥n de energ√≠a de membrana\n","membrane = pyrosetta.create_score_function(\"membrane_highres_Menv_smooth\")\n","#Espec√≠ficamos el prefijo, n√∫mero de trabajos y la funci√≥n de energ√≠a a usar\n","jd = pyrosetta.toolbox.py_jobdistributor.PyJobDistributor(pdb_name=\"GYPA_gorilla_\",nstruct=2,scorefxn=membrane)\n","#Le entregamos la pose previamente creada al distribuidor de trabajos\n","jd.native_pose = pose\n","#Creamos un dataframe de panda donde se almacenar√°n los resultados\n","gdf = pd.DataFrame()\n","#Hacemos un bucle hasta que se completen todos los trabajos\n","while not jd.job_complete:\n","    test_pose = pose.clone()\n","    xml.apply(test_pose)\n","    test_gdf = pd.DataFrame.from_records(dict(test_pose.scores), index=[jd.current_name])\n","    gdf = gdf.append(test_gdf)\n","    jd.output_decoy(test_pose)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Mh_oZyC4XgY6"},"source":["#### Ornitorrinchus ü•ö"]},{"cell_type":"code","metadata":{"id":"PmDCmCogbbja"},"source":["flags = f\"\"\"\n","-no_fconfig\n","-in:file:fasta ./input_files/fasta/ornithorhynchusGYPA.fasta\n","-out:prefix GYPA_ornithorhynchus_\n","-in:file:spanfile ./input_files/octopus/ornithorhynchus.span\n","-membrane:no_interpolate_Mpair\n","-membrane:Menv_penalties\n","-rg_reweight .1\n","-relax:fast\n","-relax:minimize_bond_angles\n","-relax:minimize_bond_lengths\n","-relax:jump_move true\n","-default_max_cycles 2\n","-relax:min_type lbfgs_armijo_nonmonotone\n","-score:weights input_files/wts/stage3_rlx_membrane.wts\n","-use_bicubic_interpolation\n","-hybridize:stage1_probability 1.0\n","-sog_upper_bound 15\n","-chemical:exclude_patches LowerDNA  UpperDNA Cterm_amidation SpecialRotamer VirtualBB ShoveBB VirtualDNAPhosphate VirtualNTerm CTermConnect sc_orbitals pro_hydroxylated_case1 pro_hydroxylated_case2 ser_phosphorylated thr_phosphorylated  tyr_phosphorylated tyr_sulfated lys_dimethylated lys_monomethylated  lys_trimethylated lys_acetylated glu_carboxylated cys_acetylated tyr_diiodinated N_acetylated C_methylamidated MethylatedProteinCterm\n","-analytic_etable_evaluation true\n","-linmem_ig 10\n","-out:mute core.io.pose_from_sfr.PoseFromSFRBuilder core.scoring.MembranePotential core.chemical.GlobalResidueTypeSet basic.io.database core.scoring.MembraneTopology core.scoring.CartesianBondedEnergy core.conformation.Conformation protocols.jd2.parser.ScoreFunctionLoader core.optimization.AtomTreeMinimizer core.pack\n","\"\"\"\n","pyrosetta.init(flags)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"NxxxpSsKbbjd"},"source":["%%time\n","pose = pose_from_sequence('IDPEIEHIFTGPVAAVIIYAVVCGVIGTILFIALVIKVV', 'fa_standard')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"XKkrl3q8bbjf"},"source":["%%time\n","xml = pyrosetta.rosetta.protocols.rosetta_scripts.XmlObjects.create_from_string(\"\"\"\n","<ROSETTASCRIPTS>\n","    <TASKOPERATIONS>\n","    </TASKOPERATIONS>\n","    <RESIDUE_SELECTORS>\n","    </RESIDUE_SELECTORS>\n","    <SCOREFXNS>\n","      <ScoreFunction name=\"stage1\" weights=\"input_files/wts/stage1_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"1\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"stage2\" weights=\"input_files/wts/stage2_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"fullatom\" weights=\"input_files/wts/stage3_rlx_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"membrane\" weights=\"membrane_highres_Menv_smooth\" symmetric=\"0\">\n","        <Reweight scoretype=\"cart_bonded\" weight=\"0.5\"/>\n","        <Reweight scoretype=\"pro_close\" weight=\"0\"/>\n","\t <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","\t</ScoreFunction>\n","    </SCOREFXNS>\n","    <FILTERS>\n","    </FILTERS>\n","    <MOVERS>\n","          AddMembraneMover name=\"add_memb\"\n","        \t <Hybridize name=\"hybridize\" stage1_scorefxn=\"stage1\" stage2_scorefxn=\"stage2\" fa_scorefxn=\"fullatom\" batch=\"1\" stage1_increase_cycles=\"0.5\" stage2_increase_cycles=\"0.5\" linmin_only=\"1\" add_hetatm=\"0\" hetatm_to_protein_cst_weight=\"0\"  >\n","       \t\t\t<Fragments three_mers=\"/content/input_files/frags/ornithorhynchusGYPA.3mers\" nine_mers=\"/content/input_files/frags/ornithorhynchusGYPA.9mers\" />\n","\t          \t<Template pdb=\"/content/input_files/threaded_pdbs/5eh6_A.ornithorhynchus.pdb\" cst_file=\"AUTO\" weight=\"1.000\" />\n","\t       </Hybridize>\n","    </MOVERS>\n","    <APPLY_TO_POSE>\n","    </APPLY_TO_POSE>\n","    <PROTOCOLS>\n","      Add mover=\"add_memb\"\n","\t    <Add mover=\"hybridize\"/>\n","    </PROTOCOLS>\n","<OUTPUT scorefxn=\"membrane\"/>\n","</ROSETTASCRIPTS>\n","    \"\"\").get_mover(\"ParsedProtocol\")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"_a7SPqrMbbji"},"source":["%%time\n","membrane = pyrosetta.create_score_function(\"membrane_highres_Menv_smooth\")\n","jd = pyrosetta.toolbox.py_jobdistributor.PyJobDistributor(pdb_name=\"GYPA_ornitorrinchus\",nstruct=2,scorefxn=membrane)\n","jd.native_pose = pose\n","odf = pd.DataFrame()\n","while not jd.job_complete:\n","    test_pose = pose.clone()\n","    xml.apply(test_pose)\n","    test_odf = pd.DataFrame.from_records(dict(test_pose.scores), index=[jd.current_name])\n","    odf = odf.append(test_odf)\n","    jd.output_decoy(test_pose)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"feibleVGXjZX"},"source":["####Marmota ü¶°"]},{"cell_type":"code","metadata":{"id":"sh2p_Qq0bee4"},"source":["%%time\n","flags = f\"\"\"\n","-no_fconfig\n","-in:file:fasta ./input_files/fasta/marmotaGYPA.fasta\n","-out:prefix GYPA_marmota_\n","-out:path:all ./output_files\n","-in:file:spanfile ./input_files/octopus/marmota.span\n","-membrane:no_interpolate_Mpair\n","-membrane:Menv_penalties\n","-rg_reweight .1\n","-relax:fast\n","-relax:minimize_bond_angles\n","-relax:minimize_bond_lengths\n","-relax:jump_move true\n","-default_max_cycles 2\n","-relax:min_type lbfgs_armijo_nonmonotone\n","-score:weights input_files/wts/stage3_rlx_membrane.wts\n","-use_bicubic_interpolation\n","-hybridize:stage1_probability 1.0\n","-sog_upper_bound 15\n","-chemical:exclude_patches LowerDNA  UpperDNA Cterm_amidation SpecialRotamer VirtualBB ShoveBB VirtualDNAPhosphate VirtualNTerm CTermConnect sc_orbitals pro_hydroxylated_case1 pro_hydroxylated_case2 ser_phosphorylated thr_phosphorylated  tyr_phosphorylated tyr_sulfated lys_dimethylated lys_monomethylated  lys_trimethylated lys_acetylated glu_carboxylated cys_acetylated tyr_diiodinated N_acetylated C_methylamidated MethylatedProteinCterm\n","-analytic_etable_evaluation true\n","-linmem_ig 10\n","-out:mute core.io.pose_from_sfr.PoseFromSFRBuilder core.scoring.MembranePotential core.chemical.GlobalResidueTypeSet basic.io.database core.scoring.MembraneTopology core.scoring.CartesianBondedEnergy core.conformation.Conformation protocols.jd2.parser.ScoreFunctionLoader core.optimization.AtomTreeMinimizer core.pack\n","\"\"\"\n","pyrosetta.init(flags)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"GahGjthKbee7"},"source":["%%time\n","pose = pose_from_sequence('KQGRNQIIHPFSEPVIILIIFAVMFGIIGTILLISFCVRRL', 'fa_standard')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"BhLwUqtBbee9"},"source":["%%time\n","xml = pyrosetta.rosetta.protocols.rosetta_scripts.XmlObjects.create_from_string(\"\"\"\n","<ROSETTASCRIPTS>\n","    <TASKOPERATIONS>\n","    </TASKOPERATIONS>\n","    <RESIDUE_SELECTORS>\n","    </RESIDUE_SELECTORS>\n","    <SCOREFXNS>\n","      <ScoreFunction name=\"stage1\" weights=\"input_files/wts/stage1_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"1\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"stage2\" weights=\"input_files/wts/stage2_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"fullatom\" weights=\"input_files/wts/stage3_rlx_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"membrane\" weights=\"membrane_highres_Menv_smooth\" symmetric=\"0\">\n","        <Reweight scoretype=\"cart_bonded\" weight=\"0.5\"/>\n","        <Reweight scoretype=\"pro_close\" weight=\"0\"/>\n","\t <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","\t</ScoreFunction>\n","    </SCOREFXNS>\n","    <FILTERS>\n","    </FILTERS>\n","    <MOVERS>\n","          AddMembraneMover name=\"add_memb\"\n","        \t <Hybridize name=\"hybridize\" stage1_scorefxn=\"stage1\" stage2_scorefxn=\"stage2\" fa_scorefxn=\"fullatom\" batch=\"1\" stage1_increase_cycles=\"0.5\" stage2_increase_cycles=\"0.5\" linmin_only=\"1\" add_hetatm=\"0\" hetatm_to_protein_cst_weight=\"0\"  >\n","       \t\t\t<Fragments three_mers=\"./input_files/frags/marmotaGYPA.3mers\" nine_mers=\"./input_files/frags/marmotaGYPA.9mers\" />\n","\t          \t<Template pdb=\"./input_files/threaded_pdbs/5eh6_A.marmota.pdb\" cst_file=\"AUTO\" weight=\"1.000\" />\n","\t       </Hybridize>\n","    </MOVERS>\n","    <APPLY_TO_POSE>\n","    </APPLY_TO_POSE>\n","    <PROTOCOLS>\n","      Add mover=\"add_memb\"\n","\t    <Add mover=\"hybridize\"/>\n","    </PROTOCOLS>\n","<OUTPUT scorefxn=\"membrane\"/>\n","</ROSETTASCRIPTS>\n","    \"\"\").get_mover(\"ParsedProtocol\")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"l2l4g1onbefA"},"source":["membrane = pyrosetta.create_score_function(\"membrane_highres_Menv_smooth\")\n","jd = pyrosetta.toolbox.py_jobdistributor.PyJobDistributor(pdb_name=\"GYPA_marmota_\",nstruct=2,scorefxn=membrane)\n","jd.native_pose = pose\n","mdf = pd.DataFrame()\n","while not jd.job_complete:\n","    test_pose = pose.clone()\n","    xml.apply(test_pose)\n","    test_mdf = pd.DataFrame.from_records(dict(test_pose.scores), index=[jd.current_name])\n","    mdf = mdf.append(test_mdf)\n","    jd.output_decoy(test_pose)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"0ehjClu0YhXk"},"source":["####Human üßë"]},{"cell_type":"code","metadata":{"id":"slUPRFWuoOQR"},"source":["%%time\n","flags = f\"\"\"\n","-no_fconfig\n","-in:file:fasta ./input_files/fasta/GYPA.fasta\n","-parser:protocol ./input_files/cm.xml\n","-out:prefix GYPA_\n","-out:path:all ./output_files\n","-in:file:spanfile ./input_files/octopus/GYPA.span\n","-membrane:no_interpolate_Mpair\n","-membrane:Menv_penalties\n","-rg_reweight .1\n","-relax:fast\n","-relax:minimize_bond_angles\n","-relax:minimize_bond_lengths\n","-relax:jump_move true\n","-default_max_cycles 2\n","-relax:min_type lbfgs_armijo_nonmonotone\n","-score:weights input_files/wts/stage3_rlx_membrane.wts\n","-use_bicubic_interpolation\n","-hybridize:stage1_probability 1.0\n","-sog_upper_bound 15\n","-chemical:exclude_patches LowerDNA  UpperDNA Cterm_amidation SpecialRotamer VirtualBB ShoveBB VirtualDNAPhosphate VirtualNTerm CTermConnect sc_orbitals pro_hydroxylated_case1 pro_hydroxylated_case2 ser_phosphorylated thr_phosphorylated  tyr_phosphorylated tyr_sulfated lys_dimethylated lys_monomethylated  lys_trimethylated lys_acetylated glu_carboxylated cys_acetylated tyr_diiodinated N_acetylated C_methylamidated MethylatedProteinCterm\n","-analytic_etable_evaluation true\n","-linmem_ig 10\n","-out:mute core.io.pose_from_sfr.PoseFromSFRBuilder core.scoring.MembranePotential core.chemical.GlobalResidueTypeSet basic.io.database core.scoring.MembraneTopology core.scoring.CartesianBondedEnergy core.conformation.Conformation protocols.jd2.parser.ScoreFunctionLoader core.optimization.AtomTreeMinimizer core.pack\n","\"\"\"\n","pyrosetta.init(flags)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"dYwsIljin-vA"},"source":["%%time\n","pose = pose_from_sequence('ETGERVQLAHHFSEPEITLIIFGVMAGVIGTILLISYGIRRL', 'fa_standard')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"oYRxZf4M8_Un"},"source":["%%time\n","xml = pyrosetta.rosetta.protocols.rosetta_scripts.XmlObjects.create_from_string(\"\"\"\n","<ROSETTASCRIPTS>\n","    <TASKOPERATIONS>\n","    </TASKOPERATIONS>\n","    <RESIDUE_SELECTORS>\n","    </RESIDUE_SELECTORS>\n","    <SCOREFXNS>\n","      <ScoreFunction name=\"stage1\" weights=\"input_files/wts/stage1_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"1\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"stage2\" weights=\"input_files/wts/stage2_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"fullatom\" weights=\"input_files/wts/stage3_rlx_membrane.wts\" symmetric=\"0\">\n","        <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","      </ScoreFunction>\n","      <ScoreFunction name=\"membrane\" weights=\"membrane_highres_Menv_smooth\" symmetric=\"0\">\n","        <Reweight scoretype=\"cart_bonded\" weight=\"0.5\"/>\n","        <Reweight scoretype=\"pro_close\" weight=\"0\"/>\n","\t <Reweight scoretype=\"atom_pair_constraint\" weight=\"0.5\"/>\n","\t</ScoreFunction>\n","    </SCOREFXNS>\n","    <FILTERS>\n","    </FILTERS>\n","    <MOVERS>\n","          AddMembraneMover name=\"add_memb\"\n","        \t <Hybridize name=\"hybridize\" stage1_scorefxn=\"stage1\" stage2_scorefxn=\"stage2\" fa_scorefxn=\"fullatom\" batch=\"1\" stage1_increase_cycles=\"0.5\" stage2_increase_cycles=\"0.5\" linmin_only=\"1\" add_hetatm=\"0\" hetatm_to_protein_cst_weight=\"0\"  >\n","       \t\t\t<Fragments three_mers=\"./input_files/frags/GYPA.3mers\" nine_mers=\"./input_files/frags/GYPA.9mers\" />\n","\t          \t<Template pdb=\"/content/input_files/threaded_pdbs/5eh6_A.human.pdb\" cst_file=\"AUTO\" weight=\"1.000\" />\n","\t       </Hybridize>\n","    </MOVERS>\n","    <APPLY_TO_POSE>\n","    </APPLY_TO_POSE>\n","    <PROTOCOLS>\n","      Add mover=\"add_memb\"\n","\t    <Add mover=\"hybridize\"/>\n","    </PROTOCOLS>\n","<OUTPUT scorefxn=\"membrane\"/>\n","</ROSETTASCRIPTS>\n","    \"\"\").get_mover(\"ParsedProtocol\")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"gvyV2upvoAW1"},"source":["%%time\n","membrane = pyrosetta.create_score_function(\"membrane_highres_Menv_smooth\")\n","jd = pyrosetta.toolbox.py_jobdistributor.PyJobDistributor(pdb_name=\"GYPA_human_\",nstruct=2,scorefxn=membrane)\n","jd.native_pose = pose\n","#Name of the score dataframe\n","hdf = pd.DataFrame()\n","while not jd.job_complete:\n","    test_pose = pose.clone()\n","    xml.apply(test_pose)\n","    test_hdf = pd.DataFrame.from_records(dict(test_pose.scores), index=[jd.current_name])\n","    hdf = hdf.append(test_hdf)\n","    jd.output_decoy(test_pose)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"1GSWo_C0qE3M"},"source":["#Parte IV ‚Äì Analizar y visualizar las estructuras modeladas"]},{"cell_type":"markdown","metadata":{"id":"eVDe0mz3QRQi"},"source":["1. Ahora, necesitamos determinar nuestros mejores modelos analizando sus DataFrames de puntuaciones. Para esto, vamos a obtener el DataFrame de cada uno de los modelos que generamos con PyRosetta.\n","\n","   Podemos inspeccionar visualmente estos DataFrames simplemente escribiendo sus nombres en un fragmento de c√≥digo.\n"]},{"cell_type":"code","metadata":{"id":"sD8nrCKprPpr"},"source":["#Invoca los diferentes dataframes que creamos\n","hdf\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"GZ1NnFdOlIT0"},"source":["odf\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"AixtSW1JlJXz"},"source":["mdf\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"yvfyIrqClKJL"},"source":["gdf"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"gKSGvDM8Okly"},"source":["2. Dado que Rosetta utiliza un algoritmo estoc√°stico para el modelado comparativo, solo estamos interesados en las estructuras de baja energ√≠a. Por lo tanto, podemos ordenar nuestros modelos por la columna total_score como se muestra en la siguiente celda de c√≥digo:\n"]},{"cell_type":"code","metadata":{"id":"0XFEzgf4Li4f"},"source":["#hdf.sort_values('total_score')\n","gdf.sort_values('total_score')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"jE8R-lrqRdXe"},"source":["3. Tener demasiadas tablas o tablas con demasiada informaci√≥n ser√≠a un desaf√≠o para analizar nuestros resultados. Para simplificar, podemos hacer un gr√°fico con el **total_score** de cada PDB para cada especie utilizando **matplotlib**.\n"]},{"cell_type":"code","metadata":{"id":"WU7m7LhCq8DG"},"source":["import matplotlib\n","import seaborn\n","frames = [hdf,odf,mdf,gdf]\n","results = pd.concat(frames)\n","matplotlib.rcParams['figure.figsize'] = [12.0, 8.0]\n","seaborn.barplot(x=\"total_score\", y=\"index\", data=results.sort_values('total_score').reset_index())"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Tc7oifPnbssg"},"source":["4. Finalmente, podemos echar un vistazo a nuestros mejores modelos utilizando **py3Dmol**.\n"]},{"cell_type":"code","metadata":{"id":"LPpEhymrJ5gM"},"source":["#!pip install py3Dmol\n","import py3Dmol\n","view=py3Dmol.view()\n","#Cargando el molde\n","view.addModel(open('5eh6.pdb', 'r').read(),'pdb')\n","view.addModel(open('GYPA_gorilla__0.pdb', 'r').read(),'pdb')\n","view.zoomTo()\n","view.setBackgroundColor('white')\n","view.setStyle({'chain':'A'},{'cartoon': {'color':'purple'}})\n","view.setStyle({'chain':' '},{'cartoon': {'color':'yellow'}})\n","view.show()\n","input()\n","view.png()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"JFpJRYPwJ8w6"},"source":["**üìöTAREA:** Como desaf√≠o, utilizando lo que has aprendido en los tutoriales anteriores para superponer estructuras y determinar su similitud mediante el RMSD, toma tus mejores estructuras y comp√°ralas para responder a la siguiente pregunta: **¬øEs diferente la regi√≥n transmembrana (TM) entre los ort√≥logos?**\n"]},{"cell_type":"markdown","metadata":{"id":"7CV7AsCQIFrB"},"source":["5. Para finalizar, puedes descargar tus archivos de salida seg√∫n se indica en la celda de c√≥digo a continuaci√≥n.\n"]},{"cell_type":"code","metadata":{"id":"oy4qyfAGMDFp"},"source":["#descarga los archivos finales\n","%%bash\n","zip -r example.zip ./outs\n","mkdir outs\n","cp *.pdb ./outs/"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"lJpeTpYPIKfB"},"source":["**Este es el final del quito tutorial. Buena ciencia!**"]},{"cell_type":"markdown","metadata":{"id":"s9giUnnEAast"},"source":["#Ap√©ndice A"]},{"cell_type":"markdown","metadata":{"id":"n1NoW-mIX3Fp"},"source":["## Generando los archivos de entrada utilizando la versi√≥n independiente de l√≠nea de comandos de Rosetta**\n"]},{"cell_type":"markdown","metadata":{"id":"FF5pgAKmh14Y"},"source":["Primero, necesitas obtener una licencia de Rosetta [AQU√ç](https://els2.comotion.uw.edu/product/rosetta), al igual que hiciste para PyRosetta."]},{"cell_type":"markdown","metadata":{"id":"pZ8OMgudYJi9"},"source":["### Descarga y extrae Rosetta en Google Colab (ETA ~15min)"]},{"cell_type":"code","metadata":{"id":"9eLMeKl_vyuf"},"source":["!wget --user=Academic_User --password=Xry3x4 https://www.rosettacommons.org/downloads/academic/3.12/rosetta_bin_linux_3.12_bundle.tgz\n","!tar xvf /content/rosetta_bin_linux_3.12_bundle.tgz"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"5XUiyaTK6nxy"},"source":["!unzip input_files.zip"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"oovXJfZUl5g7"},"source":["### Gerenar un archivo Grishin"]},{"cell_type":"markdown","metadata":{"id":"ZaQNZszVl8bf"},"source":["El formato de archivo de alineaci√≥n Grishin est√° documentado [aqu√≠](https://www.rosettacommons.org/docs/latest/rosetta_basics/file_types/Grishan-format-alignment). A continuaci√≥n, se muestra un ejemplo de su estructura:\n","\n","```\n","## target_name template_name.pdb\n","#\n","scores from program: 0\n","0 HERE_GOES_YOUR_TARGET_SEQUENCE_ALIGNMENT_IN_ONE_LINE\n","0 HERE_GOES_YOUR_TEMPLATE_SEQUENCE_ALIGNMENT_IN_ONE_LINE\n","```\n","Este es uno de los archivos Grishin reales usados en el tutorial\n","```\n","## gorillaGYPA 5eh6_A.pdb\n","#\n","scores from program: 0\n","0 DNGEWIQLVPRFSGPEITLIIFGVMAGIIGTILLISYSIRRL\n","0 --------------PEITLIIFGVIAGVIGTILLISYGIRRL\n","```\n","\n"]},{"cell_type":"code","metadata":{"id":"RmZi9ejLmvYF"},"source":["#Modifica esta celda para hacer un archivo Grishin a partir de un alineamiento clustal\n","%%bash\n","echo \\\n","\"\"\"\\\n","## gorillaGYPA 5eh6_A.pdb\n","#\n","scores from program: 0\n","0 DNGEWIQLVPRFSGPEITLIIFGVMAGIIGTILLISYSIRRL\n","0 --------------PEITLIIFGVIAGVIGTILLISYGIRRL\n","\"\"\" > /content/input_files/grishin/5eh6_on_gorillaGYPA.grishin\n","\n","echo \\\n","\"\"\"\\\n","## GYPA 5eh6_A.pdb\n","#\n","scores from program: 0\n","0 ETGERVQLAHHFSEPEITLIIFGVMAGVIGTILLISYGIRRL\n","0 --------------PEITLIIFGVIAGVIGTILLISYGIRRL\n","\"\"\" > /content/input_files/grishin/5eh6_on_humanGYPA.grishin\n","\n","echo \\\n","\"\"\"\\\n","## marmotaGYPA 5eh6_A.pdb\n","#\n","scores from program: 0\n","0 -KQGRNQIIHPFSEPVIILIIFAVMFGIIGTILLISFCVRRL\n","0 --------------PEITLIIFGVIAGVIGTILLISYGIRRL\n","\"\"\" > /content/input_files/grishin/5eh6_on_marmotaGYPA.grishin\n","\n","\n","echo \\\n","\"\"\"\\\n","## OrnithorhynchusGYPA 5eh6_A.pdb\n","#\n","scores from program: 0\n","0 ---IDPEIEHIFTGPVAAVIIYAVVCGVIGTILFIALVIKVV\n","0 --------------PEITLIIFGVIAGVIGTILLISYGIRRL\n","\"\"\" > /content/input_files/grishin/5eh6_on_ornithorhynchusGYPA.grishin"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"iAfdLyUYYV40"},"source":["### Enhebrado parcial"]},{"cell_type":"code","metadata":{"id":"COXQteMfi0Ku"},"source":["import urllib.request\n","urllib.request.urlretrieve('http://files.rcsb.org/download/5eh6.pdb', '/content/input_files/pdbs/5eh6.pdb')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"z6Z-k7TDigtJ"},"source":["%%bash\n","cd /content/input_files/pdbs\n","/content/rosetta_bin_linux_2020.08.61146_bundle/tools/protein_tools/scripts/clean_pdb.py 5eh6.pdb A\n","mv 5eh6_A.* ../cleaned"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"Yw46R_v1560X"},"source":["!/content/rosetta_bin_linux_2020.08.61146_bundle/main/source/bin/partial_thread.static.linuxgccrelease -in:file:fasta ./fasta/GYPA.fasta -in:file:alignment ./grishin/5eh6.grishin -in:file:template_pdb ./cleaned/5eh6_A.pdb\n","!mv 5eh6_A.pdb.pdb 5eh6_A.human.pdb\n","!/content/rosetta_bin_linux_2020.08.61146_bundle/main/source/bin/partial_thread.static.linuxgccrelease -in:file:fasta ./fasta/ornithorhynchusGYPA.fasta -in:file:alignment ./grishin/5eh6_on_ornithorhynchusGYPA.grishin -in:file:template_pdb ./cleaned/5eh6_A.pdb\n","!mv 5eh6_A.pdb.pdb 5eh6_A.ornithorhynchus.pdb\n","!/content/rosetta_bin_linux_2020.08.61146_bundle/main/source/bin/partial_thread.static.linuxgccrelease -in:file:fasta ./fasta/marmotaGYPA.fasta -in:file:alignment ./grishin/5eh6_on_marmotaGYPA.grishin -in:file:template_pdb ./cleaned/5eh6_A.pdb\n","!mv 5eh6_A.pdb.pdb 5eh6_A.gorilla.pdb\n","!/content/rosetta_bin_linux_2020.08.61146_bundle/main/source/bin/partial_thread.static.linuxgccrelease -in:file:fasta ./fasta/gorillaGYPA.fasta -in:file:alignment ./grishin/5eh6_on_gorillaGYPA.grishin -in:file:template_pdb ./cleaned/5eh6_A.pdb\n","!mv 5eh6_A.pdb.pdb 5eh6_A.marmota.pdb\n","!mv 5eh6_A.*pdb ./input_files/threaded_pdbs/"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"2Wn-5LuU44Wm"},"source":["### Obteniendo fragmentos desde oldRobetta server\n"]},{"cell_type":"markdown","metadata":{"id":"LmPktANT5uEe"},"source":["Seg√∫n la documentaci√≥n de Rosetta, los fragmentos son:\n","\n","> (...) utilizados en el ensamblaje de prote√≠nas, ya sea para la predicci√≥n de la estructura o el dise√±o, para reducir el tama√±o del espacio de b√∫squeda para el plegamiento de prote√≠nas. Son una parte central del dise√±o de Rosetta.\n","\n","Para el prop√≥sito de este tutorial, hemos omitido esta parte. Sin embargo, en la siguiente figura puedes ver una captura de pantalla del [servidor Robetta](http://old.robetta.org/fragmentsubmit.jsp) donde puedes obtener los fragmentos. Para ello, debes registrarte y enviar una secuencia FASTA de tus prote√≠nas objetivo. Ten en cuenta que la longitud de tu objetivo afectar√° el tiempo necesario para el proceso de selecci√≥n de fragmentos.\n"]},{"cell_type":"markdown","metadata":{"id":"Neo5VCZVJoBl"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/mpmodelling_04.png'/>\n","</figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"xvXqKZZdZNjJ"},"source":["Cuando tu trabajo haya terminado, recibir√°s un correo electr√≥nico con los fragmentos correspondientes. Debes descargarlos y colocarlos dentro de la carpeta \"frags\", asegur√°ndote de que el nombre de archivo coincida con el especificado en la celda de RosettaScripts XML.\n"]}]}